// react-gauge-component/dist/lib/GaugeComponent/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import React, { useEffect, useRef, useLayoutEffect, useState } from "react";
import { createPortal } from "react-dom";
import { pie } from "d3-shape";
import { select } from "d3-selection";
import { defaultGaugeProps, getGaugeMarginByType } from "./types/GaugeComponentProps";
import * as chartHooks from "./hooks/chart";
import * as arcHooks from "./hooks/arc";
import { mergeObjects } from "./hooks/utils";
import { defaultDimensions } from "./types/Dimensions";
import { defaultPointerRef, getPointerLengthByType } from "./types/Pointer";
import { getArcWidthByType } from "./types/Arc";
import CONSTANTS from "./constants";
/**
 * GaugeComponent - A responsive gauge chart component built with D3
 *
 * Features:
 * - Responsive design that adapts to container size
 * - Multiple gauge types: semicircle, radial, grafana
 * - Customizable arcs, pointers, labels, and tooltips
 * - Smooth animations with configurable timing
 * - ResizeObserver for automatic resize handling
 */
// CSS keyframes for fade-in animation to prevent initial render flash
var fadeInKeyframes = "\n@keyframes gaugeComponentFadeIn {\n  from { opacity: 0; }\n  to { opacity: 1; }\n}\n";
// Inject keyframes once
if (typeof document !== 'undefined' && !document.getElementById('gauge-fade-in-style')) {
    var style = document.createElement('style');
    style.id = 'gauge-fade-in-style';
    style.textContent = fadeInKeyframes;
    document.head.appendChild(style);
}
var GaugeComponent = function (props) {
    var svg = useRef({});
    var tooltip = useRef({});
    var g = useRef({});
    var doughnut = useRef({});
    var isFirstRun = useRef(true);
    var currentProgress = useRef(0);
    var pointer = useRef(__assign({}, defaultPointerRef));
    var container = useRef({});
    var arcData = useRef([]);
    var parentNode = useRef();
    var pieChart = useRef(pie());
    var dimensions = useRef(__assign(__assign({}, defaultDimensions), { margin: __assign({}, defaultDimensions.margin), angles: __assign({}, defaultDimensions.angles) }));
    var mergedProps = useRef(props);
    var prevProps = useRef({});
    var prevGSize = useRef(null);
    var maxGHeight = useRef(null);
    var svgRef = useRef(null);
    var customContent = useRef({});
    var initialAnimationTriggered = useRef(false);
    var animationInProgress = useRef(false);
    var pendingResize = useRef(false);
    var pendingConfigChange = useRef(false); // Track if config changed during animation
    var multiPointers = useRef([]);
    var multiPointerAnimationTriggered = useRef([]);
    var isDragging = useRef(false); // Track if user is dragging pointer
    var hasBeenInitialized = useRef(false); // Track if component has ever been initialized
    var measuredBoundsRef = useRef(null); // Persist measured bounds across renders
    var renderPassRef = useRef(1); // Persist render pass state
    var initialRenderDeferred = useRef(false); // Track if we're deferring initial render to ResizeObserver
    var lastContainerSize = useRef(null); // Track last container dimensions to prevent infinite recalculation
    // State to trigger re-render when custom content needs to be rendered
    var _a = useState([]), customContentItems = _a[0], setCustomContentItems = _a[1];
    // Helper function to sync custom content state after chart render
    // This needs to be called after initChart/renderChart to update React state
    // with any custom content that was registered during the D3 render pass
    var syncCustomContentState = function () {
        var customContentConfig = customContent.current;
        if (Array.isArray(customContentConfig === null || customContentConfig === void 0 ? void 0 : customContentConfig.items) && customContentConfig.items.length > 0) {
            // Always update with the new items since DOM nodes change on each render
            // We can't use JSON.stringify because items contain functions and DOM nodes
            setCustomContentItems(__spreadArray([], customContentConfig.items, true));
        }
        else if ((customContentConfig === null || customContentConfig === void 0 ? void 0 : customContentConfig.domNode) && (customContentConfig === null || customContentConfig === void 0 ? void 0 : customContentConfig.renderContent)) {
            // Backward compatible single-item config
            setCustomContentItems([{
                    domNode: customContentConfig.domNode,
                    renderContent: customContentConfig.renderContent,
                    value: customContentConfig.value,
                    arcColor: customContentConfig.arcColor,
                }]);
        }
        else {
            setCustomContentItems(function (prev) { return prev.length > 0 ? [] : prev; });
        }
    };
    // Store sync function in a ref so ResizeObserver callback always has access to latest version
    var syncCustomContentStateRef = useRef(syncCustomContentState);
    syncCustomContentStateRef.current = syncCustomContentState;
    // Use a ref for gauge so the ResizeObserver always has access to current props
    var gaugeRef = useRef(null);
    var gauge = {
        props: mergedProps.current,
        originalProps: props,
        prevProps: prevProps,
        resizeObserver: useRef(),
        svg: svg,
        g: g,
        dimensions: dimensions,
        doughnut: doughnut,
        isFirstRun: isFirstRun,
        currentProgress: currentProgress,
        pointer: pointer,
        container: container,
        arcData: arcData,
        pieChart: pieChart,
        tooltip: tooltip,
        prevGSize: prevGSize,
        maxGHeight: maxGHeight,
        customContent: customContent,
        initialAnimationTriggered: initialAnimationTriggered,
        animationInProgress: animationInProgress,
        pendingResize: pendingResize,
        pendingConfigChange: pendingConfigChange,
        multiPointers: multiPointers,
        multiPointerAnimationTriggered: multiPointerAnimationTriggered,
        isDragging: isDragging,
        // Persisted refs for two-pass rendering stability
        measuredBounds: measuredBoundsRef,
        renderPass: renderPassRef,
    };
    // Keep gaugeRef updated with current gauge (including current props)
    gaugeRef.current = gauge;
    // Merge default props with user-provided props
    var updateMergedProps = function () {
        var _a, _b, _c;
        var defaultValues = __assign({}, defaultGaugeProps);
        gauge.props = mergedProps.current = mergeObjects(defaultValues, props);
        // Apply type-specific defaults
        if (((_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.width) === ((_b = defaultGaugeProps.arc) === null || _b === void 0 ? void 0 : _b.width)) {
            var mergedArc = mergedProps.current.arc;
            mergedArc.width = getArcWidthByType(gauge.props.type);
        }
        if (gauge.props.marginInPercent === defaultGaugeProps.marginInPercent) {
            mergedProps.current.marginInPercent = getGaugeMarginByType(gauge.props.type);
        }
        // Apply type-specific pointer length: if user didn't explicitly set length,
        // override the generic default (0.70) with the type-specific default
        if (((_c = props.pointer) === null || _c === void 0 ? void 0 : _c.length) === undefined) {
            var mergedPointer = mergedProps.current.pointer;
            if (mergedPointer === null || mergedPointer === void 0 ? void 0 : mergedPointer.type) {
                mergedPointer.length = getPointerLengthByType(mergedPointer.type);
            }
        }
        arcHooks.validateArcs(gauge);
    };
    // Determine if chart should be re-initialized based on prop changes
    // IMPORTANT: Only trigger reinit for STRUCTURAL changes
    // Value changes should NOT trigger reinit - they just animate the pointer
    var shouldInitChart = function () {
        var _a, _b;
        var prev = prevProps.current;
        var curr = mergedProps.current;
        // Structural changes that require full reinit
        var arcsPropsChanged = JSON.stringify(prev.arc) !== JSON.stringify(curr.arc);
        // For labels, exclude functions and internal properties from comparison
        // __renderIndex is added internally and should not trigger reinit
        var getLabelsStructure = function (l) {
            if (!l)
                return null;
            // Clean ticks array - remove internal __renderIndex property
            var cleanTicks = function (ticks) {
                if (!ticks)
                    return undefined;
                return ticks.map(function (t) {
                    var __renderIndex = t.__renderIndex, rest = __rest(t, ["__renderIndex"]);
                    return rest;
                });
            };
            return {
                valueLabel: l.valueLabel ? {
                    hide: l.valueLabel.hide,
                    matchColorWithArc: l.valueLabel.matchColorWithArc,
                    maxDecimalDigits: l.valueLabel.maxDecimalDigits,
                    style: l.valueLabel.style,
                    animateValue: l.valueLabel.animateValue,
                } : undefined,
                tickLabels: l.tickLabels ? {
                    hideMinMax: l.tickLabels.hideMinMax,
                    type: l.tickLabels.type,
                    ticks: cleanTicks(l.tickLabels.ticks),
                    defaultTickValueConfig: l.tickLabels.defaultTickValueConfig ? {
                        hide: l.tickLabels.defaultTickValueConfig.hide,
                        maxDecimalDigits: l.tickLabels.defaultTickValueConfig.maxDecimalDigits,
                        style: l.tickLabels.defaultTickValueConfig.style,
                    } : undefined,
                    defaultTickLineConfig: l.tickLabels.defaultTickLineConfig,
                } : undefined,
            };
        };
        var labelsPropsChanged = JSON.stringify(getLabelsStructure(prev.labels)) !== JSON.stringify(getLabelsStructure(curr.labels));
        var typeChanged = prev.type !== curr.type;
        var minValueChanged = prev.minValue !== curr.minValue;
        var maxValueChanged = prev.maxValue !== curr.maxValue;
        var anglesChanged = prev.startAngle !== curr.startAngle || prev.endAngle !== curr.endAngle;
        // For single pointer, only check if hide changed (show/hide requires reinit)
        // Other prop changes (type, color, length, width, etc.) are handled by renderChart
        var prevPointer = prev.pointer || {};
        var currPointer = curr.pointer || {};
        var pointerHideChanged = prevPointer.hide !== currPointer.hide;
        // For pointers array, only check COUNT changes (structure/prop changes handled by renderChart)
        // This prevents full reinit when just changing pointer type/color/etc.
        var prevPointers = prev.pointers;
        var currPointers = curr.pointers;
        var pointersCountChanged = ((_a = prevPointers === null || prevPointers === void 0 ? void 0 : prevPointers.length) !== null && _a !== void 0 ? _a : 0) !== ((_b = currPointers === null || currPointers === void 0 ? void 0 : currPointers.length) !== null && _b !== void 0 ? _b : 0);
        // Mode transition (single <-> multi pointer) requires reinit
        var wasMultiPointer = Array.isArray(prevPointers) && prevPointers.length > 0;
        var isMultiPointer = Array.isArray(currPointers) && currPointers.length > 0;
        var modeTransition = wasMultiPointer !== isMultiPointer;
        return arcsPropsChanged || labelsPropsChanged || pointerHideChanged || pointersCountChanged ||
            modeTransition || typeChanged || minValueChanged || maxValueChanged || anglesChanged;
    };
    // Check if only value changed (for animation without reinit)
    var onlyValueChanged = function () {
        var _a;
        var prev = prevProps.current;
        var curr = mergedProps.current;
        return prev.value !== curr.value ||
            ((_a = prev.pointers) === null || _a === void 0 ? void 0 : _a.some(function (p, i) { var _a, _b; return p.value !== ((_b = (_a = curr.pointers) === null || _a === void 0 ? void 0 : _a[i]) === null || _b === void 0 ? void 0 : _b.value); }));
    };
    var isHeightProvidedByUser = function () { var _a; return ((_a = mergedProps.current.style) === null || _a === void 0 ? void 0 : _a.height) !== undefined; };
    var isHeightPresentInParentNode = function () { var _a; return ((_a = parentNode.current) === null || _a === void 0 ? void 0 : _a.clientHeight) !== 0; };
    // Initialize and update chart on prop changes
    useLayoutEffect(function () {
        updateMergedProps();
        // Use dedicated flag instead of checking container content
        // Container can appear "empty" after removing old content, but it's not first render
        var isFirstRender = !hasBeenInitialized.current;
        isFirstRun.current = isFirstRender;
        if (CONSTANTS.debugLogs) {
            //console.debug("isHeightProvidedByUser:", isHeightProvidedByUser());
            //console.debug("isHeightPresentInParentNode:", isHeightPresentInParentNode());
        }
        if (isFirstRender) {
            container.current = select(svgRef.current);
            hasBeenInitialized.current = true;
            // MOBILE FIX: Defer initial render to ResizeObserver callback
            // On mobile devices, getBoundingClientRect() can return incorrect dimensions
            // during the initial render because the browser hasn't finished layout.
            // By deferring to ResizeObserver, we ensure dimensions are accurate.
            // ResizeObserver fires immediately when observing an element with stable dimensions.
            initialRenderDeferred.current = true;
            gauge.prevProps.current = mergedProps.current;
            return; // Skip initial chart render - let ResizeObserver handle it
        }
        // Check if structural changes require full reinit
        var needsReinit = shouldInitChart();
        var valueOnlyChange = onlyValueChanged();
        if (needsReinit) {
            chartHooks.initChart(gauge, isFirstRender);
        }
        else {
            // Any non-structural change (value, animation settings, etc.) - just re-render
            chartHooks.renderChart(gauge, false);
        }
        gauge.prevProps.current = mergedProps.current;
        // Sync custom content state after chart render
        syncCustomContentState();
    }, [props]);
    // Set up ResizeObserver for responsive resizing
    useEffect(function () {
        var element = svgRef.current;
        if (!element)
            return;
        var resizeTimeout;
        var fallbackTimeout;
        var handleResize = function (entries) {
            // Cancel any pending resize to debounce rapid changes
            if (resizeTimeout) {
                clearTimeout(resizeTimeout);
            }
            var entry = entries[0];
            if (!entry)
                return;
            var newWidth = entry.contentRect.width;
            var newHeight = entry.contentRect.height;
            // Skip if dimensions are invalid (but don't skip deferred initial render fallback)
            if (newWidth <= 0 || newHeight <= 0) {
                // If initial render is deferred and we got zero dimensions,
                // the fallback timeout will handle it
                return;
            }
            // Clear fallback timeout since we got valid dimensions
            if (fallbackTimeout) {
                clearTimeout(fallbackTimeout);
            }
            // INFINITE LOOP FIX: Skip if container dimensions haven't actually changed
            // This prevents oscillation caused by internal viewBox recalculations
            // triggering ResizeObserver which triggers more recalculations
            var lastSize = lastContainerSize.current;
            if (lastSize && !initialRenderDeferred.current) {
                var widthChange = Math.abs(newWidth - lastSize.width);
                var heightChange = Math.abs(newHeight - lastSize.height);
                // Use 1px threshold to ignore sub-pixel changes
                if (widthChange < 1 && heightChange < 1) {
                    return;
                }
            }
            // Update last known container size
            lastContainerSize.current = { width: newWidth, height: newHeight };
            // Log resize event for debugging
            if (CONSTANTS.debugLogs) {
                // console.log('[ResizeObserver] Element resized:', {
                //   width: newWidth,
                //   height: newHeight
                // });
            }
            // Use a small delay to ensure layout is stable
            resizeTimeout = setTimeout(function () {
                requestAnimationFrame(function () {
                    // Use gaugeRef.current to always get the latest gauge with current props
                    if (gaugeRef.current) {
                        // MOBILE FIX: Handle deferred initial render
                        // If this is the first resize after mount, perform the initial chart render
                        // This ensures we render with accurate dimensions from ResizeObserver
                        if (initialRenderDeferred.current) {
                            initialRenderDeferred.current = false;
                            chartHooks.initChart(gaugeRef.current, true);
                            // CRITICAL: Sync custom content state after deferred initial render
                            // This ensures renderContent labels are displayed on first render
                            syncCustomContentStateRef.current();
                        }
                        else {
                            chartHooks.renderChart(gaugeRef.current, true);
                            // Sync custom content state after resize render
                            syncCustomContentStateRef.current();
                        }
                    }
                });
            }, 16); // ~1 frame
        };
        var observer = new ResizeObserver(handleResize);
        // Observe the gauge container itself, not its parent
        // This ensures we get accurate dimensions when we resize
        observer.observe(element);
        // FALLBACK: If ResizeObserver doesn't provide valid dimensions within 100ms,
        // render anyway using getBoundingClientRect as fallback
        // This handles cases where container has height:100% but parent has no explicit height
        if (initialRenderDeferred.current) {
            fallbackTimeout = setTimeout(function () {
                if (initialRenderDeferred.current && gaugeRef.current) {
                    initialRenderDeferred.current = false;
                    chartHooks.initChart(gaugeRef.current, true);
                    // CRITICAL: Sync custom content state after fallback initial render
                    syncCustomContentStateRef.current();
                }
            }, 100);
        }
        return function () {
            if (resizeTimeout) {
                clearTimeout(resizeTimeout);
            }
            if (fallbackTimeout) {
                clearTimeout(fallbackTimeout);
            }
            observer.disconnect();
        };
    }, []);
    var id = props.id, style = props.style, className = props.className;
    // Container must properly fill its parent and not overflow
    var containerStyle = __assign({ width: '100%', height: '100%', maxWidth: '100%', maxHeight: '100%', overflow: 'hidden' }, style);
    // Render custom content via React portal if configured
    var renderCustomContent = function () {
        if (!customContentItems || customContentItems.length === 0)
            return null;
        return (React.createElement(React.Fragment, null, customContentItems
            .filter(function (it) { return (it === null || it === void 0 ? void 0 : it.domNode) && (it === null || it === void 0 ? void 0 : it.renderContent); })
            .map(function (it, idx) {
            return createPortal(it.renderContent(it.value, it.arcColor), it.domNode, "gauge-custom-content-".concat(idx));
        })));
    };
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { id: id, className: "gauge".concat(className ? ' ' + className : ''), style: containerStyle, ref: function (ref) { return (svgRef.current = ref); } }),
        renderCustomContent()));
};
export default GaugeComponent;
