// react-gauge-component/dist/lib/GaugeComponent/types/GaugeComponentProps.js
import { defaultArc } from "./Arc";
import { defaultLabels } from './Labels';
import { defaultPointer } from "./Pointer";
export var GaugeType;
(function (GaugeType) {
    GaugeType["Semicircle"] = "semicircle";
    GaugeType["Radial"] = "radial";
    GaugeType["Grafana"] = "grafana";
})(GaugeType || (GaugeType = {}));
export var defaultGaugeProps = {
    id: Math.random().toString(36).slice(2),
    className: "gauge-component-class",
    style: { width: "100%" },
    marginInPercent: 0.07,
    value: 33,
    minValue: 0,
    maxValue: 100,
    arc: defaultArc,
    labels: defaultLabels,
    pointer: defaultPointer,
    type: GaugeType.Grafana,
    fadeInAnimation: false
};
export var getGaugeMarginByType = function (type) {
    var _a;
    var gaugeTypesMargin = (_a = {},
        _a[GaugeType.Grafana] = { top: 0.12, bottom: 0.00, left: 0.07, right: 0.07 },
        _a[GaugeType.Semicircle] = { top: 0.08, bottom: 0.00, left: 0.08, right: 0.08 },
        _a[GaugeType.Radial] = { top: 0.07, bottom: 0.00, left: 0.07, right: 0.07 },
        _a);
    return gaugeTypesMargin[type];
};
