// react-gauge-component/dist/lib/GaugeComponent/types/Pointer.js
export var PointerType;
(function (PointerType) {
    PointerType["Needle"] = "needle";
    PointerType["Blob"] = "blob";
    PointerType["Arrow"] = "arrow";
})(PointerType || (PointerType = {}));
export var defaultPointerContext = {
    centerPoint: [0, 0],
    pointerRadius: 0,
    pathLength: 0,
    currentPercent: 0,
    prevPercent: 0,
    prevProgress: 0,
    pathStr: "",
    shouldDrawPath: false,
    prevColor: ""
};
export var defaultPointerRef = {
    element: undefined,
    path: undefined,
    context: defaultPointerContext
};
export var defaultPointer = {
    type: PointerType.Needle,
    color: undefined,
    baseColor: "white",
    length: 0.70,
    width: 20,
    animate: true,
    elastic: false,
    hide: false,
    animationDuration: 3000,
    animationDelay: 100,
    strokeWidth: 0,
    strokeColor: undefined,
    arrowOffset: 0.72,
    blobOffset: 0.5,
    hideGrabHandle: true,
    // Performance defaults - 60fps, fine threshold
    maxFps: 60,
    animationThreshold: 0.001
};
/**
 * Returns the default pointer length for a given pointer type.
 * Arrow pointers are shorter (0.2) since they're positioned on the arc,
 * while Needle pointers extend from center (0.7).
 * Blob pointers don't use pathLength (positioned by blobOffset), so length is irrelevant.
 */
export var getPointerLengthByType = function (type) {
    var _a;
    var _b;
    var pointerTypesLength = (_a = {},
        _a[PointerType.Needle] = 0.70,
        _a[PointerType.Arrow] = 0.20,
        _a[PointerType.Blob] = 0.70,
        _a);
    if (!type)
        type = PointerType.Needle;
    return (_b = pointerTypesLength[type]) !== null && _b !== void 0 ? _b : 0.70;
};
