// react-gauge-component/dist/lib/GaugeComponent/hooks/animationCoordinator.js
/**
 * Animation Coordinator - Global animation loop manager for all gauge components
 *
 * This module provides a centralized animation system that:
 * - Batches animations across all gauge instances into a single RAF loop
 * - Prevents visual stuttering by distributing updates
 * - Allows FPS limiting at the global level
 * - Supports animation priority and staggering
 */
var AnimationCoordinator = /** @class */ (function () {
    function AnimationCoordinator() {
        var _this = this;
        this.animations = new Map();
        this.isRunning = false;
        this.lastFrameTime = 0;
        this.frameId = null;
        this.targetFps = 60;
        this.minFrameTime = 1000 / 60;
        this.tick = function () {
            if (!_this.isRunning)
                return;
            var now = performance.now();
            var deltaTime = now - _this.lastFrameTime;
            // FPS limiting
            if (deltaTime < _this.minFrameTime) {
                _this.frameId = requestAnimationFrame(_this.tick);
                return;
            }
            _this.lastFrameTime = now;
            // Sort animations by priority (higher first)
            var sortedAnimations = Array.from(_this.animations.values())
                .sort(function (a, b) { return b.priority - a.priority; });
            // Process animations
            var completedIds = [];
            for (var _i = 0, sortedAnimations_1 = sortedAnimations; _i < sortedAnimations_1.length; _i++) {
                var entry = sortedAnimations_1[_i];
                try {
                    var isComplete = !entry.callback(now, deltaTime);
                    if (isComplete) {
                        completedIds.push(entry.id);
                    }
                }
                catch (error) {
                    console.error("[AnimationCoordinator] Error in animation ".concat(entry.id, ":"), error);
                    completedIds.push(entry.id);
                }
            }
            // Remove completed animations
            for (var _a = 0, completedIds_1 = completedIds; _a < completedIds_1.length; _a++) {
                var id = completedIds_1[_a];
                _this.animations.delete(id);
            }
            // Continue loop if there are active animations
            if (_this.animations.size > 0) {
                _this.frameId = requestAnimationFrame(_this.tick);
            }
            else {
                _this.stop();
            }
        };
    }
    AnimationCoordinator.getInstance = function () {
        if (!AnimationCoordinator.instance) {
            AnimationCoordinator.instance = new AnimationCoordinator();
        }
        return AnimationCoordinator.instance;
    };
    /**
     * Set the target FPS for all animations
     */
    AnimationCoordinator.prototype.setTargetFps = function (fps) {
        this.targetFps = Math.max(1, Math.min(120, fps));
        this.minFrameTime = 1000 / this.targetFps;
    };
    /**
     * Register an animation callback
     * @param id Unique identifier for this animation
     * @param callback Function called each frame, returns false when complete
     * @param priority Higher priority animations are updated first (default: 0)
     */
    AnimationCoordinator.prototype.register = function (id, callback, priority) {
        if (priority === void 0) { priority = 0; }
        this.animations.set(id, {
            id: id,
            callback: callback,
            priority: priority,
            startTime: performance.now()
        });
        if (!this.isRunning) {
            this.start();
        }
    };
    /**
     * Unregister an animation
     */
    AnimationCoordinator.prototype.unregister = function (id) {
        this.animations.delete(id);
        if (this.animations.size === 0) {
            this.stop();
        }
    };
    /**
     * Check if an animation is currently registered
     */
    AnimationCoordinator.prototype.isRegistered = function (id) {
        return this.animations.has(id);
    };
    /**
     * Get the number of active animations
     */
    AnimationCoordinator.prototype.getActiveCount = function () {
        return this.animations.size;
    };
    AnimationCoordinator.prototype.start = function () {
        if (this.isRunning)
            return;
        this.isRunning = true;
        this.lastFrameTime = performance.now();
        this.tick();
    };
    AnimationCoordinator.prototype.stop = function () {
        this.isRunning = false;
        if (this.frameId !== null) {
            cancelAnimationFrame(this.frameId);
            this.frameId = null;
        }
    };
    /**
     * Stagger animation starts to prevent all gauges from animating at exact same time
     * Returns a delay in ms based on current animation count
     */
    AnimationCoordinator.prototype.getStaggerDelay = function (baseDelay) {
        if (baseDelay === void 0) { baseDelay = 0; }
        var count = this.animations.size;
        // Stagger by 16ms (one frame) per active animation, up to 100ms max
        var stagger = Math.min(count * 16, 100);
        return baseDelay + stagger;
    };
    // Singleton instance
    AnimationCoordinator.instance = null;
    return AnimationCoordinator;
}());
// Export singleton instance
export var animationCoordinator = AnimationCoordinator.getInstance();
// Export utility functions
export var registerAnimation = function (id, callback, priority) {
    animationCoordinator.register(id, callback, priority);
};
export var unregisterAnimation = function (id) {
    animationCoordinator.unregister(id);
};
export var setGlobalAnimationFps = function (fps) {
    animationCoordinator.setTargetFps(fps);
};
export var getAnimationStaggerDelay = function (baseDelay) {
    if (baseDelay === void 0) { baseDelay = 0; }
    return animationCoordinator.getStaggerDelay(baseDelay);
};
export var isAnimationActive = function (id) {
    return animationCoordinator.isRegistered(id);
};
export var getActiveAnimationCount = function () {
    return animationCoordinator.getActiveCount();
};
