// react-gauge-component/dist/lib/GaugeComponent/hooks/arc.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import * as utils from './utils';
import { select } from "d3-selection";
import { scaleLinear } from "d3-scale";
import { interpolateHsl } from "d3-interpolate";
import { arc } from "d3-shape";
import * as arcHooks from './arc';
import CONSTANTS from '../constants';
import { defaultTooltipStyle } from '../types/Tooltip';
import { GaugeType } from '../types/GaugeComponentProps';
import { throttle } from './throttle';
/**
 * Get effective angles for a gauge, with fallback to type defaults if not yet calculated.
 * This ensures ticks are positioned correctly even on first render before dimensions are set.
 */
export var getEffectiveAngles = function (gauge) {
    var _a = gauge.dimensions.current.angles, startAngle = _a.startAngle, endAngle = _a.endAngle;
    // If angles are set (non-zero or explicitly custom), use them
    if (startAngle !== 0 || endAngle !== 0) {
        return { startAngle: startAngle, endAngle: endAngle };
    }
    // Check for custom angles in props
    if (gauge.props.startAngle !== undefined && gauge.props.endAngle !== undefined) {
        return {
            startAngle: (gauge.props.startAngle * Math.PI) / 180,
            endAngle: (gauge.props.endAngle * Math.PI) / 180
        };
    }
    // Fallback to type defaults
    switch (gauge.props.type) {
        case GaugeType.Semicircle:
            return { startAngle: -Math.PI / 2 + 0.02, endAngle: Math.PI / 2 - 0.02 };
        case GaugeType.Radial:
            return { startAngle: -Math.PI / 1.37, endAngle: Math.PI / 1.37 };
        case GaugeType.Grafana:
        default:
            return { startAngle: -Math.PI / 1.6, endAngle: Math.PI / 1.6 };
    }
};
var onArcMouseMove = function (event, d, gauge) {
    //event.target.style.stroke = "#ffffff5e";
    if (d.data.tooltip != undefined) {
        var shouldChangeText = d.data.tooltip.text != gauge.tooltip.current.text();
        if (shouldChangeText) {
            gauge.tooltip.current.html(d.data.tooltip.text)
                .style("position", "absolute")
                .style("display", "block")
                .style("opacity", 1);
            applyTooltipStyles(d.data.tooltip, d.data.color, gauge);
        }
        gauge.tooltip.current.style("left", (event.pageX + 15) + "px")
            .style("top", (event.pageY - 10) + "px");
    }
    if (d.data.onMouseMove != undefined)
        d.data.onMouseMove(event);
};
var applyTooltipStyles = function (tooltip, arcColor, gauge) {
    //Apply default styles
    Object.entries(defaultTooltipStyle).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return gauge.tooltip.current.style(utils.camelCaseToKebabCase(key), value);
    });
    gauge.tooltip.current.style("background-color", arcColor);
    //Apply custom styles
    if (tooltip.style != undefined)
        Object.entries(tooltip.style).forEach(function (_a) {
            var key = _a[0], value = _a[1];
            return gauge.tooltip.current.style(utils.camelCaseToKebabCase(key), value);
        });
};
var onArcMouseLeave = function (event, d, gauge, mousemoveCbThrottled) {
    mousemoveCbThrottled.cancel();
    hideTooltip(gauge);
    if (d.data.onMouseLeave != undefined)
        d.data.onMouseLeave(event);
};
export var hideTooltip = function (gauge) {
    gauge.tooltip.current.html(" ").style("display", "none");
};
var onArcMouseOut = function (event, d, gauge) {
    var _a, _b, _c, _e;
    // Restore configured stroke instead of removing it
    var strokeWidth = ((_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.subArcsStrokeWidth) || 0;
    if (strokeWidth > 0) {
        var strokeColor = ((_e = (_c = gauge.props) === null || _c === void 0 ? void 0 : _c.arc) === null || _e === void 0 ? void 0 : _e.subArcsStrokeColor) || '#ffffff';
        event.target.style.stroke = strokeColor;
        event.target.style.strokeWidth = strokeWidth;
    }
    else {
        event.target.style.stroke = "none";
    }
};
var onArcMouseClick = function (event, d) {
    if (d.data.onMouseClick != undefined)
        d.data.onMouseClick(event);
};
export var setArcData = function (gauge) {
    var _a, _b;
    var arc = gauge.props.arc;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    // Determine number of arcs to display
    var nbArcsToDisplay = (arc === null || arc === void 0 ? void 0 : arc.nbSubArcs) || (((_a = arc === null || arc === void 0 ? void 0 : arc.subArcs) === null || _a === void 0 ? void 0 : _a.length) || 1);
    var colorArray = getColors(nbArcsToDisplay, gauge);
    if ((arc === null || arc === void 0 ? void 0 : arc.subArcs) && !(arc === null || arc === void 0 ? void 0 : arc.nbSubArcs)) {
        var lastSubArcLimit_1 = 0;
        var lastSubArcLimitPercentageAcc_1 = 0;
        var subArcsLength_1 = [];
        var subArcsLimits_1 = [];
        var subArcsTooltip_1 = [];
        (_b = arc === null || arc === void 0 ? void 0 : arc.subArcs) === null || _b === void 0 ? void 0 : _b.forEach(function (subArc, i) {
            var _a;
            var subArcLength = 0;
            //map limit for non defined subArcs limits
            var subArcRange = 0;
            var limit = subArc.limit;
            if (subArc.length != undefined) {
                subArcLength = subArc.length;
                limit = utils.getCurrentGaugeValueByPercentage(subArcLength + lastSubArcLimitPercentageAcc_1, gauge);
            }
            else if (subArc.limit == undefined) {
                subArcRange = lastSubArcLimit_1;
                var remainingSubArcs = (_a = arc === null || arc === void 0 ? void 0 : arc.subArcs) === null || _a === void 0 ? void 0 : _a.slice(i);
                var remainingPercentage = 1 - utils.calculatePercentage(minValue, maxValue, lastSubArcLimit_1);
                var remainingPercentageEquallyDivided = remainingPercentage / Math.max((remainingSubArcs === null || remainingSubArcs === void 0 ? void 0 : remainingSubArcs.length) || 1, 1);
                // Convert from percentage space to value space correctly
                limit = utils.getCurrentGaugeValueByPercentage(lastSubArcLimitPercentageAcc_1 + remainingPercentageEquallyDivided, gauge);
                subArcLength = remainingPercentageEquallyDivided;
            }
            else {
                subArcRange = limit - lastSubArcLimit_1;
                // Calculate arc length based on previous arc percentage
                if (i !== 0) {
                    subArcLength = utils.calculatePercentage(minValue, maxValue, limit) - lastSubArcLimitPercentageAcc_1;
                }
                else {
                    subArcLength = utils.calculatePercentage(minValue, maxValue, subArcRange);
                }
            }
            subArcsLength_1.push(subArcLength);
            subArcsLimits_1.push(limit);
            lastSubArcLimitPercentageAcc_1 = subArcsLength_1.reduce(function (count, curr) { return count + curr; }, 0);
            lastSubArcLimit_1 = limit;
            if (subArc.tooltip != undefined)
                subArcsTooltip_1.push(subArc.tooltip);
        });
        var subArcs_1 = arc.subArcs;
        gauge.arcData.current = subArcsLength_1.map(function (length, i) { return ({
            value: length,
            limit: subArcsLimits_1[i],
            color: colorArray[i],
            showTick: subArcs_1[i].showTick || false,
            tooltip: subArcs_1[i].tooltip || undefined,
            onMouseMove: subArcs_1[i].onMouseMove,
            onMouseLeave: subArcs_1[i].onMouseLeave,
            onMouseClick: subArcs_1[i].onClick
        }); });
    }
    else {
        // Calculate arc value based on the RANGE, not maxValue
        var range = maxValue - minValue;
        var arcValue_1 = range / nbArcsToDisplay;
        gauge.arcData.current = Array.from({ length: nbArcsToDisplay }, function (_, i) { return ({
            value: arcValue_1,
            limit: minValue + (i + 1) * arcValue_1,
            color: colorArray[i],
            tooltip: undefined,
        }); });
    }
};
var getGrafanaMainArcData = function (gauge, percent) {
    var _a;
    if (percent === void 0) { percent = undefined; }
    // Check if multi-pointer mode with subarcs between pointers
    var pointers = gauge.props.pointers;
    if (pointers && pointers.length > 1) {
        return getGrafanaMultiPointerArcData(gauge);
    }
    var currentPercentage = percent != undefined ? percent : utils.calculatePercentage(gauge.props.minValue, gauge.props.maxValue, gauge.props.value);
    var curArcData = getArcDataByPercentage(currentPercentage, gauge);
    var firstSubArc = {
        value: currentPercentage,
        //White indicate that no arc was found and work as an alert for debug
        color: (curArcData === null || curArcData === void 0 ? void 0 : curArcData.color) || "white",
        //disabled for now because onMouseOut is not working properly with the
        //high amount of renderings of this arc
        //tooltip: curArcData?.tooltip
    };
    //This is the grey arc that will be displayed when the gauge is not full
    var secondSubArc = {
        value: 1 - currentPercentage,
        color: ((_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.emptyColor) || '#3a3a3a',
    };
    return [firstSubArc, secondSubArc];
};
/**
 * Generate arc data for multi-pointer Grafana mode.
 * Creates subarcs between each pointer, colored with the next pointer's color.
 *
 * Example with 3 pointers at values 20, 50, 80:
 * - Arc 1 (0-20): color of pointer 1
 * - Arc 2 (20-50): color of pointer 2
 * - Arc 3 (50-80): color of pointer 3
 * - Arc 4 (80-100): empty color
 */
var getGrafanaMultiPointerArcData = function (gauge) {
    var _a, _b;
    var pointers = gauge.props.pointers;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var emptyColor = ((_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.emptyColor) || '#3a3a3a';
    // Sort pointers by value
    var sortedPointers = __spreadArray([], pointers, true).sort(function (a, b) { return a.value - b.value; });
    // Build arc segments
    var arcData = [];
    var prevPercent = 0;
    for (var i = 0; i < sortedPointers.length; i++) {
        var pointer = sortedPointers[i];
        var currentPercent = utils.calculatePercentage(minValue, maxValue, pointer.value);
        // Arc from previous position to this pointer
        var segmentSize = currentPercent - prevPercent;
        if (segmentSize > 0) {
            // Use this pointer's color for the segment leading up to it
            var arcColor = (_b = getArcDataByPercentage(currentPercent, gauge)) === null || _b === void 0 ? void 0 : _b.color;
            var color = pointer.color || (typeof arcColor === 'string' ? arcColor : '#888');
            arcData.push({
                value: segmentSize,
                color: color
            });
        }
        prevPercent = currentPercent;
    }
    // Add remaining empty arc (from last pointer to 100%)
    var remainingPercent = 1 - prevPercent;
    if (remainingPercent > 0) {
        arcData.push({
            value: remainingPercent,
            color: emptyColor
        });
    }
    return arcData;
};
var drawGrafanaOuterArc = function (gauge, resize) {
    var _a, _b, _c;
    if (resize === void 0) { resize = false; }
    var outerRadius = gauge.dimensions.current.outerRadius;
    var _e = gauge.props.arc, padding = _e.padding, cornerRadius = _e.cornerRadius, outerArcConfig = _e.outerArc;
    //Grafana's outer arc will be populated as the standard arc data would
    if (gauge.props.type == GaugeType.Grafana && resize) {
        gauge.doughnut.current.selectAll(".outerSubArc").remove();
        // Use outerArc config if provided, otherwise fall back to main arc settings
        // Get outer arc width (default 5px)
        var outerArcWidth = (_a = outerArcConfig === null || outerArcConfig === void 0 ? void 0 : outerArcConfig.width) !== null && _a !== void 0 ? _a : 5;
        // Scale corner radius for the thin outer arc (max 2px since arc is only 5px thick)
        var maxCornerRadius = Math.min(outerArcWidth / 2, 2);
        var outerCornerRadius = (outerArcConfig === null || outerArcConfig === void 0 ? void 0 : outerArcConfig.cornerRadius) !== undefined
            ? Math.min(outerArcConfig.cornerRadius, maxCornerRadius)
            : (cornerRadius ? Math.min(cornerRadius, maxCornerRadius) : 0);
        // Outer arc has its own padding - does NOT inherit from main arc
        var outerPadding = (_b = outerArcConfig === null || outerArcConfig === void 0 ? void 0 : outerArcConfig.padding) !== null && _b !== void 0 ? _b : 0;
        var padEndpoints = (_c = gauge.props.arc) === null || _c === void 0 ? void 0 : _c.padEndpoints;
        // Gap between main arc and outer arc is 2px
        var gap = 2;
        var outerArc = arc()
            .outerRadius(outerRadius + gap + outerArcWidth)
            .innerRadius(outerRadius + gap)
            .cornerRadius(outerCornerRadius)
            .padAngle(padEndpoints === false ? 0 : outerPadding);
        // Generate pie data
        var outerPieData_1 = gauge.pieChart.current(gauge.arcData.current);
        // Apply manual padding when padEndpoints is false
        if (padEndpoints === false && outerPadding > 0 && outerPieData_1.length > 1) {
            var halfPad_1 = outerPadding / 2;
            outerPieData_1 = outerPieData_1.map(function (d, i) {
                var adjusted = __assign({}, d);
                if (i < outerPieData_1.length - 1)
                    adjusted.endAngle -= halfPad_1;
                if (i > 0)
                    adjusted.startAngle += halfPad_1;
                return adjusted;
            });
        }
        var arcPaths = gauge.doughnut.current
            .selectAll("anyString")
            .data(outerPieData_1)
            .enter()
            .append("g")
            .attr("class", "outerSubArc");
        var outerArcSubarcs = arcPaths
            .append("path")
            .attr("d", outerArc);
        applyColors(outerArcSubarcs, gauge);
        var mousemoveCbThrottled_1 = throttle(function (event, d) { return onArcMouseMove(event, d, gauge); }, 20);
        arcPaths
            .on("mouseleave", function (event, d) { return onArcMouseLeave(event, d, gauge, mousemoveCbThrottled_1); })
            .on("mouseout", function (event, d) { return onArcMouseOut(event, d, gauge); })
            .on("mousemove", mousemoveCbThrottled_1)
            .on("click", function (event, d) { return onArcMouseClick(event, d); });
    }
};
export var drawArc = function (gauge, percent) {
    var _a, _b, _c;
    if (percent === void 0) { percent = undefined; }
    var _e = gauge.props.arc, padding = _e.padding, cornerRadius = _e.cornerRadius;
    var _f = gauge.dimensions.current, innerRadius = _f.innerRadius, outerRadius = _f.outerRadius;
    var data = {};
    //When gradient enabled, it'll have only 1 arc
    if ((_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.gradient) {
        data = [{ value: 1 }];
    }
    else {
        data = gauge.arcData.current;
    }
    if (gauge.props.type == GaugeType.Grafana) {
        data = getGrafanaMainArcData(gauge, percent);
    }
    var arcPadding = padding || 0;
    var arcCornerRadius = cornerRadius || 0;
    var padEndpoints = (_c = gauge.props.arc) === null || _c === void 0 ? void 0 : _c.padEndpoints;
    // When padEndpoints is false, we handle padding manually via adjusted angles
    // so we don't use padAngle on the arc generator
    var arcObj = arc()
        .outerRadius(outerRadius)
        .innerRadius(innerRadius)
        .cornerRadius(arcCornerRadius)
        .padAngle(padEndpoints === false ? 0 : arcPadding);
    // Generate pie data and adjust for endpoint padding
    var pieData = gauge.pieChart.current(data);
    // Apply manual padding to pie data when padEndpoints is false
    if (padEndpoints === false && arcPadding > 0 && pieData.length > 1) {
        var halfPad_2 = arcPadding / 2;
        pieData = pieData.map(function (d, i) {
            var adjusted = __assign({}, d);
            // Add padding to end of each arc (except the last one)
            if (i < pieData.length - 1) {
                adjusted.endAngle -= halfPad_2;
            }
            // Add padding to start of each arc (except the first one)
            if (i > 0) {
                adjusted.startAngle += halfPad_2;
            }
            return adjusted;
        });
    }
    var arcPaths = gauge.doughnut.current
        .selectAll("anyString")
        .data(pieData)
        .enter()
        .append("g")
        .attr("class", "subArc");
    var subArcs = arcPaths
        .append("path")
        .attr("d", arcObj);
    applyColors(subArcs, gauge);
    var mousemoveCbThrottled = throttle(function (event, d) { return onArcMouseMove(event, d, gauge); }, 20);
    arcPaths
        .on("mouseleave", function (event, d) { return onArcMouseLeave(event, d, gauge, mousemoveCbThrottled); })
        .on("mouseout", function (event, d) { return onArcMouseOut(event, d, gauge); })
        .on("mousemove", mousemoveCbThrottled)
        .on("click", function (event, d) { return onArcMouseClick(event, d); });
};
export var setupArcs = function (gauge, resize, initialPercent) {
    if (resize === void 0) { resize = false; }
    if (initialPercent === void 0) { initialPercent = undefined; }
    //Setup the arc
    setupTooltip(gauge);
    drawGrafanaOuterArc(gauge, resize);
    // For Grafana type with animation, pass initialPercent (usually 0 or prevPercent) 
    // so the arc starts at the correct position before animating
    drawArc(gauge, initialPercent);
};
export var setupTooltip = function (gauge) {
    //Add tooltip
    var isTooltipInTheDom = document.getElementsByClassName(CONSTANTS.arcTooltipClassname).length != 0;
    if (!isTooltipInTheDom)
        select("body").append("div").attr("class", CONSTANTS.arcTooltipClassname);
    gauge.tooltip.current = select(".".concat(CONSTANTS.arcTooltipClassname));
    gauge.tooltip.current
        .on("mouseleave", function () { return arcHooks.hideTooltip(gauge); })
        .on("mouseout", function () { return arcHooks.hideTooltip(gauge); });
};
export var applyColors = function (subArcsPath, gauge) {
    var _a, _b, _c, _e, _f, _g;
    if ((_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.gradient) {
        var uniqueId_1 = "subArc-linear-gradient-".concat(Math.random());
        var gradEl = createGradientElement(gauge.doughnut.current, uniqueId_1);
        applyGradientColors(gradEl, gauge);
        subArcsPath.style("fill", function (d) { return "url(#".concat(uniqueId_1, ")"); });
    }
    else {
        subArcsPath.style("fill", function (d) { return d.data.color; });
    }
    // Apply stroke/border to subArcs if configured
    var strokeWidth = ((_e = (_c = gauge.props) === null || _c === void 0 ? void 0 : _c.arc) === null || _e === void 0 ? void 0 : _e.subArcsStrokeWidth) || 0;
    if (strokeWidth > 0) {
        var strokeColor = ((_g = (_f = gauge.props) === null || _f === void 0 ? void 0 : _f.arc) === null || _g === void 0 ? void 0 : _g.subArcsStrokeColor) || '#ffffff';
        subArcsPath
            .attr("stroke", strokeColor)
            .attr("stroke-width", strokeWidth);
    }
    // Apply visual effects if configured
    applyArcEffects(subArcsPath, gauge);
};
/**
 * Create and apply SVG filter effects to arcs
 */
export var applyArcEffects = function (subArcsPath, gauge) {
    var _a, _b, _c, _e, _f, _g, _h, _j, _k;
    var effects = (_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.effects;
    if (!effects)
        return;
    var filterId = "arc-effects-".concat(Math.random().toString(36).substr(2, 9));
    // Check if we need to create any filters
    var needsFilter = effects.glow || effects.dropShadow || effects.innerShadow || effects.filterUrl;
    if (!needsFilter)
        return;
    // Use custom filter URL if provided
    if (effects.filterUrl) {
        subArcsPath.attr("filter", effects.filterUrl);
        return;
    }
    // Create defs if it doesn't exist
    var defs = gauge.doughnut.current.select("defs");
    if (defs.empty()) {
        defs = gauge.doughnut.current.append("defs");
    }
    // Create filter element
    var filter = defs.append("filter")
        .attr("id", filterId)
        .attr("x", "-50%")
        .attr("y", "-50%")
        .attr("width", "200%")
        .attr("height", "200%");
    var lastResult = "SourceGraphic";
    // Add glow effect
    if (effects.glow) {
        var glowBlur = (_c = effects.glowBlur) !== null && _c !== void 0 ? _c : 10;
        var glowSpread = (_e = effects.glowSpread) !== null && _e !== void 0 ? _e : 3;
        // Create glow using Gaussian blur
        filter.append("feGaussianBlur")
            .attr("in", "SourceGraphic")
            .attr("stdDeviation", glowBlur)
            .attr("result", "blur");
        // Brighten the blur for more intensity
        filter.append("feComponentTransfer")
            .attr("in", "blur")
            .attr("result", "glow")
            .selectAll("feFuncA")
            .data([null])
            .enter()
            .append("feFuncA")
            .attr("type", "linear")
            .attr("slope", glowSpread);
        // Merge glow behind original
        var merge = filter.append("feMerge");
        merge.append("feMergeNode").attr("in", "glow");
        merge.append("feMergeNode").attr("in", "SourceGraphic");
        lastResult = "glow";
    }
    // Add drop shadow effect
    if (effects.dropShadow) {
        var ds = effects.dropShadow;
        var dx = (_f = ds.dx) !== null && _f !== void 0 ? _f : 0;
        var dy = (_g = ds.dy) !== null && _g !== void 0 ? _g : 2;
        var blur_1 = (_h = ds.blur) !== null && _h !== void 0 ? _h : 3;
        var opacity = (_j = ds.opacity) !== null && _j !== void 0 ? _j : 0.3;
        var color = (_k = ds.color) !== null && _k !== void 0 ? _k : "rgba(0,0,0,".concat(opacity, ")");
        filter.append("feDropShadow")
            .attr("dx", dx)
            .attr("dy", dy)
            .attr("stdDeviation", blur_1)
            .attr("flood-color", color)
            .attr("flood-opacity", opacity);
    }
    // Add inner shadow for 3D effect
    if (effects.innerShadow) {
        // Create inner shadow using morphology and blur
        filter.append("feMorphology")
            .attr("in", "SourceAlpha")
            .attr("operator", "erode")
            .attr("radius", "1")
            .attr("result", "eroded");
        filter.append("feGaussianBlur")
            .attr("in", "eroded")
            .attr("stdDeviation", "2")
            .attr("result", "innerBlur");
        filter.append("feComposite")
            .attr("in", "innerBlur")
            .attr("in2", "SourceAlpha")
            .attr("operator", "arithmetic")
            .attr("k2", "-1")
            .attr("k3", "1")
            .attr("result", "innerShadow");
        var merge = filter.append("feMerge");
        merge.append("feMergeNode").attr("in", "SourceGraphic");
        merge.append("feMergeNode").attr("in", "innerShadow");
    }
    // Apply the filter to the arcs
    subArcsPath.attr("filter", "url(#".concat(filterId, ")"));
};
export var getArcDataByValue = function (value, gauge) {
    return gauge.arcData.current.find(function (subArcData) { return value <= subArcData.limit; });
};
export var getArcDataByPercentage = function (percentage, gauge) {
    return getArcDataByValue(utils.getCurrentGaugeValueByPercentage(percentage, gauge), gauge);
};
/**
 * Get the interpolated color for a given percentage when using gradient mode.
 * This is needed because gradient mode doesn't store colors in arcData.current.
 * Falls back to getArcDataByPercentage for non-gradient mode.
 */
export var getColorByPercentage = function (percentage, gauge) {
    var _a, _b, _c, _e, _f, _g;
    // For non-gradient mode, use the existing arc data lookup
    if (!((_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.gradient)) {
        var arcData = getArcDataByPercentage(percentage, gauge);
        return (arcData === null || arcData === void 0 ? void 0 : arcData.color) || '#fff';
    }
    // For gradient mode, interpolate color from subArcs
    var subArcs = (_c = gauge.props.arc) === null || _c === void 0 ? void 0 : _c.subArcs;
    if (!subArcs || subArcs.length === 0) {
        return '#fff';
    }
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var value = utils.getCurrentGaugeValueByPercentage(percentage, gauge);
    // Find which subArc segment the value falls into
    var prevLimit = minValue;
    for (var i = 0; i < subArcs.length; i++) {
        var subArc = subArcs[i];
        var limit = (_e = subArc.limit) !== null && _e !== void 0 ? _e : maxValue;
        if (value <= limit) {
            // Value is in this segment
            if (i === 0) {
                return subArc.color || '#fff';
            }
            // Interpolate between previous and current color
            var prevColor = ((_f = subArcs[i - 1]) === null || _f === void 0 ? void 0 : _f.color) || '#fff';
            var currentColor = subArc.color || '#fff';
            var segmentPercentage = (value - prevLimit) / (limit - prevLimit);
            // Use d3 color interpolation
            var colorInterpolator = interpolateHsl(prevColor, currentColor);
            return colorInterpolator(segmentPercentage);
        }
        prevLimit = limit;
    }
    // Value exceeds all limits, return last color
    return ((_g = subArcs[subArcs.length - 1]) === null || _g === void 0 ? void 0 : _g.color) || '#fff';
};
export var applyGradientColors = function (gradEl, gauge) {
    gauge.arcData.current.forEach(function (subArcData) {
        var _a, _b, _c, _e;
        var normalizedOffset = utils.normalize(subArcData === null || subArcData === void 0 ? void 0 : subArcData.limit, (_b = (_a = gauge === null || gauge === void 0 ? void 0 : gauge.props) === null || _a === void 0 ? void 0 : _a.minValue) !== null && _b !== void 0 ? _b : 0, (_e = (_c = gauge === null || gauge === void 0 ? void 0 : gauge.props) === null || _c === void 0 ? void 0 : _c.maxValue) !== null && _e !== void 0 ? _e : 100);
        gradEl.append("stop")
            .attr("offset", "".concat(normalizedOffset, "%"))
            .style("stop-color", subArcData.color) //end in red
            .style("stop-opacity", 1);
    });
};
//Depending on the number of levels in the chart
//This function returns the same number of colors
export var getColors = function (nbArcsToDisplay, gauge) {
    var _a;
    var arc = gauge.props.arc;
    var colorsValue = [];
    if (!arc.colorArray) {
        var subArcColors = (_a = arc.subArcs) === null || _a === void 0 ? void 0 : _a.map(function (subArc) { return subArc.color; });
        colorsValue = (subArcColors === null || subArcColors === void 0 ? void 0 : subArcColors.some(function (color) { return color != undefined; })) ? subArcColors : CONSTANTS.defaultColors;
    }
    else {
        colorsValue = arc.colorArray;
    }
    //defaults colorsValue to white in order to avoid compilation error
    if (!colorsValue)
        colorsValue = ["#fff"];
    //Check if the number of colors equals the number of levels
    //Otherwise make an interpolation
    var arcsEqualsColorsLength = nbArcsToDisplay === (colorsValue === null || colorsValue === void 0 ? void 0 : colorsValue.length);
    if (arcsEqualsColorsLength)
        return colorsValue;
    // Create multi-stop color scale that interpolates through ALL colors
    var numColors = colorsValue.length;
    // Create domain points for each color (evenly distributed from 1 to nbArcsToDisplay)
    var domainPoints = colorsValue.map(function (_, i) {
        return 1 + (i * (nbArcsToDisplay - 1)) / (numColors - 1);
    });
    var colorScale = scaleLinear()
        .domain(domainPoints)
        //@ts-ignore
        .range(colorsValue) // Use ALL colors in the range
        //@ts-ignore
        .interpolate(interpolateHsl);
    var colorArray = [];
    for (var i = 1; i <= nbArcsToDisplay; i++) {
        colorArray.push(colorScale(i));
    }
    return colorArray;
};
export var createGradientElement = function (div, uniqueId) {
    //make defs and add the linear gradient
    var lg = div.append("defs").append("linearGradient")
        .attr("id", uniqueId) //id of the gradient
        .attr("x1", "0%")
        .attr("x2", "100%")
        .attr("y1", "0%")
        .attr("y2", "0%");
    return lg;
};
export var getCoordByValue = function (value, gauge, position, centerToArcLengthSubtract) {
    var _a, _b, _c;
    if (position === void 0) { position = "inner"; }
    if (centerToArcLengthSubtract === void 0) { centerToArcLengthSubtract = 0; }
    var _e = gauge.dimensions.current, outerRadius = _e.outerRadius, innerRadius = _e.innerRadius;
    var isGrafana = gauge.props.type === GaugeType.Grafana;
    // Grafana has an outer decorative arc - get width from config (default 5px) + 2px gap
    var outerArcWidth = (_c = (_b = (_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.outerArc) === null || _b === void 0 ? void 0 : _b.width) !== null && _c !== void 0 ? _c : 5;
    var grafanaOuterArcEdge = outerRadius + 2 + outerArcWidth;
    var positionCenterToArcLength = {
        "outer": function () {
            // For outer ticks: start at the outer edge of the arc
            // Grafana: outer decorative arc edge (outerRadius + 7)
            // Others: main arc outer edge (outerRadius)
            var arcEdge = isGrafana ? grafanaOuterArcEdge : outerRadius;
            return arcEdge - centerToArcLengthSubtract;
        },
        "inner": function () {
            // For inner ticks: start at the inner edge of the arc
            // centerToArcLengthSubtract is positive to push inward (toward center)
            return innerRadius - centerToArcLengthSubtract;
        },
        "between": function () {
            var lengthBetweenOuterAndInner = (outerRadius - innerRadius);
            var middlePosition = innerRadius + lengthBetweenOuterAndInner / 2;
            return middlePosition;
        }
    };
    var centerToArcLength = positionCenterToArcLength[position]();
    var percent = utils.calculatePercentage(gauge.props.minValue, gauge.props.maxValue, value);
    // Use actual angles from gauge dimensions (supports custom angles)
    // D3 angles: 0 = top (12 o'clock), positive = clockwise
    // Math angles for tick positioning: 0 = right (3 o'clock), 90 = down
    // Conversion: mathAngle = d3Angle + π/2
    var _f = getEffectiveAngles(gauge), d3Start = _f.startAngle, d3End = _f.endAngle;
    var d3Angle = d3Start + percent * (d3End - d3Start);
    var angle = d3Angle + Math.PI / 2;
    // Calculate position relative to center (0, 0)
    // Since g is now centered at gaugeCenter, coordinates are relative to origin
    var x = -centerToArcLength * Math.cos(angle);
    var y = -centerToArcLength * Math.sin(angle);
    return { x: x, y: y };
};
export var redrawArcs = function (gauge) {
    clearArcs(gauge);
    setArcData(gauge);
    setupArcs(gauge);
};
export var clearArcs = function (gauge) {
    var _a;
    // Safety check - doughnut might not be initialized on mobile during deferred render
    if (!((_a = gauge.doughnut.current) === null || _a === void 0 ? void 0 : _a.selectAll))
        return;
    gauge.doughnut.current.selectAll(".subArc").remove();
};
/**
 * Efficiently update Grafana arc for animation frames.
 * Instead of clearing and recreating DOM elements on every frame,
 * this updates the existing arc paths' data, which is MUCH faster.
 */
export var updateGrafanaArc = function (gauge, percent) {
    var _a;
    var _b = gauge.dimensions.current, innerRadius = _b.innerRadius, outerRadius = _b.outerRadius;
    var _c = gauge.props.arc, padding = _c.padding, cornerRadius = _c.cornerRadius;
    var padEndpoints = (_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.padEndpoints;
    var arcPadding = padding || 0;
    // Get the new arc data
    var data = getGrafanaMainArcData(gauge, percent);
    // Create the arc generator with config values for corner and padding
    var arcObj = arc()
        .outerRadius(outerRadius)
        .innerRadius(innerRadius)
        .cornerRadius(cornerRadius || 0)
        .padAngle(padEndpoints === false ? 0 : arcPadding);
    // Get existing arc groups
    var existingArcGroups = gauge.doughnut.current.selectAll(".subArc");
    // For multi-pointer mode, the number of segments can change - redraw if mismatch
    var existingCount = existingArcGroups.size();
    if (existingArcGroups.empty() || existingCount !== data.length) {
        // Clear and redraw arcs when segment count changes
        gauge.doughnut.current.selectAll(".subArc").remove();
        drawArc(gauge, percent);
        return;
    }
    // Generate pie data
    var pieData = gauge.pieChart.current(data);
    // Apply manual padding when padEndpoints is false
    if (padEndpoints === false && arcPadding > 0 && pieData.length > 1) {
        var halfPad_3 = arcPadding / 2;
        pieData = pieData.map(function (d, i) {
            var adjusted = __assign({}, d);
            if (i < pieData.length - 1)
                adjusted.endAngle -= halfPad_3;
            if (i > 0)
                adjusted.startAngle += halfPad_3;
            return adjusted;
        });
    }
    // Update each arc group's path with the corresponding pie data
    existingArcGroups.each(function (_d, i) {
        var _a, _b, _c, _e;
        if (i < pieData.length) {
            var pathEl = select(this).select("path")
                .datum(pieData[i])
                .attr("d", arcObj)
                .style("fill", pieData[i].data.color);
            // Apply stroke if configured
            var strokeWidth = ((_b = (_a = gauge.props) === null || _a === void 0 ? void 0 : _a.arc) === null || _b === void 0 ? void 0 : _b.subArcsStrokeWidth) || 0;
            if (strokeWidth > 0) {
                var strokeColor = ((_e = (_c = gauge.props) === null || _c === void 0 ? void 0 : _c.arc) === null || _e === void 0 ? void 0 : _e.subArcsStrokeColor) || '#ffffff';
                pathEl.attr("stroke", strokeColor).attr("stroke-width", strokeWidth);
            }
        }
    });
};
export var clearOuterArcs = function (gauge) {
    gauge.doughnut.current.selectAll(".outerSubArc").remove();
};
export var validateArcs = function (gauge) {
    // Debug: log when validation is triggered
    // if (typeof window !== 'undefined' && (window as any).__GAUGE_DEBUG__) {
    //   console.log('[GaugeComponent] validateArcs called:', {
    //     minValue: gauge.props.minValue,
    //     maxValue: gauge.props.maxValue,
    //     gradient: gauge.props.arc?.gradient,
    //     subArcsCount: gauge.props.arc?.subArcs?.length,
    //     subArcLimits: gauge.props.arc?.subArcs?.map(s => s.limit)
    //   });
    // }
    verifySubArcsLimits(gauge);
};
/**
 * Reorders the subArcs within the gauge's arc property based on the limit property.
 * SubArcs with undefined limits are sorted last.
*/
var reOrderSubArcs = function (gauge) {
    var _a;
    var subArcs = (_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.subArcs;
    subArcs.sort(function (a, b) {
        if (typeof a.limit === 'undefined' && typeof b.limit === 'undefined') {
            return 0;
        }
        if (typeof a.limit === 'undefined') {
            return 1;
        }
        if (typeof b.limit === 'undefined') {
            return -1;
        }
        return a.limit - b.limit;
    });
};
var verifySubArcsLimits = function (gauge) {
    var _a, _b;
    // disabled when length implemented.
    // reOrderSubArcs(gauge);
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var arc = gauge.props.arc;
    var subArcs = arc.subArcs;
    var prevLimit = undefined;
    // Skip validation if using gradient mode (limits don't apply)
    if ((_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.gradient)
        return;
    // Helper to create debug info string
    var getDebugInfo = function () {
        var _a, _b, _c;
        try {
            return JSON.stringify({
                minValue: minValue,
                maxValue: maxValue,
                value: gauge.props.value,
                gradient: (_a = gauge.props.arc) === null || _a === void 0 ? void 0 : _a.gradient,
                type: gauge.props.type,
                subArcs: (_c = (_b = gauge.props.arc) === null || _b === void 0 ? void 0 : _b.subArcs) === null || _c === void 0 ? void 0 : _c.map(function (s) { return ({
                    limit: s.limit,
                    length: s.length,
                    color: typeof s.color === 'string' ? s.color.substring(0, 10) : s.color
                }); })
            }, null, 2);
        }
        catch (e) {
            return 'Unable to stringify gauge props';
        }
    };
    for (var _i = 0, _c = ((_b = gauge.props.arc) === null || _b === void 0 ? void 0 : _b.subArcs) || []; _i < _c.length; _i++) {
        var subArc = _c[_i];
        var limit = subArc.limit;
        // Only validate if limit is explicitly defined (skip length-based subArcs)
        if (typeof limit !== 'undefined') {
            // Check if the limit is within the valid range - clamp instead of throwing
            if (limit < minValue || limit > maxValue) {
                console.warn("[GaugeComponent] SubArc limit ".concat(limit, " is outside range [").concat(minValue, ", ").concat(maxValue, "]. ") +
                    "Clamping to valid range. Consider updating your subArc limits to match minValue/maxValue.");
                // Clamp the limit to valid range
                subArc.limit = Math.max(minValue, Math.min(maxValue, limit));
            }
            // Check if the limit is greater than the previous limit
            if (typeof prevLimit !== 'undefined') {
                if (subArc.limit !== undefined && subArc.limit <= prevLimit) {
                    console.warn("[GaugeComponent] SubArc limit ".concat(subArc.limit, " must be > previous limit ").concat(prevLimit, ". ") +
                        "Adjusting automatically.");
                    // Adjust to be slightly greater than previous
                    subArc.limit = prevLimit + (maxValue - minValue) * 0.01;
                }
            }
            prevLimit = subArc.limit;
        }
    }
    // If the user has defined subArcs with limits, make sure the last one covers maxValue
    // Only apply to subArcs using limit (not length)
    if (subArcs && subArcs.length > 0) {
        var lastSubArc = subArcs[subArcs.length - 1];
        // Only set limit if this subArc uses limits (not length-based)
        if (typeof lastSubArc.limit !== 'undefined' && lastSubArc.limit < maxValue) {
            lastSubArc.limit = maxValue;
        }
    }
};
