// react-gauge-component/dist/lib/GaugeComponent/hooks/coordinateSystem.js
var _a;
import { GaugeType } from "../types/GaugeComponentProps";
/**
 * Default padding config for when no labels config is provided
 */
var DEFAULT_PADDING_CONFIG = {
    hasOuterTicks: false,
    hasInnerTicks: false,
    hasTickLabels: false,
    tickLabelFontSize: 12,
    hasValueLabel: true,
    valueLabelFontSize: 35,
    pointerLength: 0.8,
};
/**
 * Extract padding configuration from gauge props
 */
export var extractPaddingConfig = function (labels, pointerLength) {
    var _a, _b, _c, _d, _e;
    var tickLabels = labels === null || labels === void 0 ? void 0 : labels.tickLabels;
    var valueLabel = labels === null || labels === void 0 ? void 0 : labels.valueLabel;
    // Determine if we have ticks and their type
    var hasTicks = tickLabels &&
        (((_a = tickLabels.ticks) === null || _a === void 0 ? void 0 : _a.length) || 0) > 0 &&
        !tickLabels.hideMinMax;
    var tickType = (tickLabels === null || tickLabels === void 0 ? void 0 : tickLabels.type) || 'outer';
    // Get font sizes from styles
    var tickFontSize = parseFloat(String(((_c = (_b = tickLabels === null || tickLabels === void 0 ? void 0 : tickLabels.defaultTickValueConfig) === null || _b === void 0 ? void 0 : _b.style) === null || _c === void 0 ? void 0 : _c.fontSize) || '12px')) || 12;
    var valueFontSize = parseFloat(String(((_d = valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.style) === null || _d === void 0 ? void 0 : _d.fontSize) || '35px')) || 35;
    return {
        hasOuterTicks: !!(hasTicks && tickType === 'outer'),
        hasInnerTicks: !!(hasTicks && tickType === 'inner'),
        hasTickLabels: !!(hasTicks && !((_e = tickLabels === null || tickLabels === void 0 ? void 0 : tickLabels.defaultTickValueConfig) === null || _e === void 0 ? void 0 : _e.hide)),
        tickLabelFontSize: tickFontSize,
        hasValueLabel: !(valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.hide),
        valueLabelFontSize: valueFontSize,
        pointerLength: pointerLength || 0.8,
    };
};
/**
 * Calculate dynamic padding based on actual gauge configuration
 * This is the key to efficient space usage!
 *
 * IMPORTANT: Tick labels at the edges (0° and 180°) extend HORIZONTALLY,
 * so side padding needs to account for text WIDTH, not just height.
 * Similarly, tick labels at top (90°) extend vertically.
 */
var calculateDynamicPadding = function (gaugeType, config, outerRadius) {
    // Calculate scale factor for tick elements (same as in labels.ts)
    // Reference radius is 100px - ticks scale proportionally
    var tickScaleFactor = Math.max(0.5, Math.min(1.5, outerRadius / 100));
    // Base padding - minimal padding for the arc itself
    // This accounts for stroke width and small visual margins
    // Scale base padding with radius for small gauges
    var basePadding = Math.max(4, outerRadius * 0.03);
    // Tick line extends outward: length (~7px) + distance from arc (~3px)
    // These are scaled proportionally with tick scale factor
    var tickLineExtension = 10 * tickScaleFactor;
    // Tick label dimensions - scale font size with gauge size
    // Text height is approximately 1.2x font size
    // Text width for numbers like "100" is approximately 2.5x font size
    var scaledFontSize = config.tickLabelFontSize * tickScaleFactor;
    var tickTextHeight = scaledFontSize * 1.3;
    var tickTextWidth = scaledFontSize * 3; // Account for "100%" type labels
    // Calculate tick padding for different directions
    var tickTopPadding = 0;
    var tickSidePadding = 0;
    if (config.hasOuterTicks) {
        if (config.hasTickLabels) {
            // Top: tick line + text height (labels at top are horizontal)
            tickTopPadding = tickLineExtension + tickTextHeight + 3 * tickScaleFactor;
            // Sides: tick line + text width (labels at 0° and 180° extend horizontally)
            tickSidePadding = tickLineExtension + tickTextWidth + 3 * tickScaleFactor;
        }
        else {
            // Just tick lines, no labels
            tickTopPadding = tickLineExtension + 2 * tickScaleFactor;
            tickSidePadding = tickLineExtension + 2 * tickScaleFactor;
        }
    }
    // Pointer padding - needles can extend beyond the arc
    var pointerPadding = config.pointerLength > 0.85
        ? outerRadius * (config.pointerLength - 0.85) * 0.5
        : 0;
    // Calculate per-direction padding
    var topPadding = basePadding + tickTopPadding + pointerPadding;
    var sidePadding = basePadding + tickSidePadding;
    var bottomPadding = basePadding;
    // Value label affects bottom padding for semicircle
    // Scale value label impact proportionally
    if (config.hasValueLabel && gaugeType === GaugeType.Semicircle) {
        // Value label sits below center, need space for it
        bottomPadding += config.valueLabelFontSize * 0.5 * tickScaleFactor;
    }
    // Grafana has an outer decorative arc that extends outward
    // Scale this padding with the gauge size
    if (gaugeType === GaugeType.Grafana) {
        topPadding += 8 * tickScaleFactor;
        sidePadding += 8 * tickScaleFactor;
    }
    // Ensure minimum padding for safety - elements should NEVER be cut off
    // Use smaller minimums for smaller gauges
    var minTopPercent = config.hasOuterTicks ? 0.15 : 0.05;
    var minSidePercent = config.hasOuterTicks ? 0.18 : 0.05; // More side padding for horizontal labels
    var minBottomPercent = 0.05;
    return {
        top: Math.max(topPadding, outerRadius * minTopPercent),
        side: Math.max(sidePadding, outerRadius * minSidePercent),
        bottom: Math.max(bottomPadding, outerRadius * minBottomPercent),
    };
};
/**
 * LEGACY: Fixed padding configuration for backward compatibility
 * Used when no labels config is available
 */
var GAUGE_TYPE_CONFIG = (_a = {},
    _a[GaugeType.Semicircle] = {
        topPaddingPercent: 0.20,
        bottomPaddingPercent: 0.08,
        sidePaddingPercent: 0.20, // Reduced from 0.38
    },
    _a[GaugeType.Radial] = {
        topPaddingPercent: 0.20,
        bottomPaddingPercent: 0.06,
        sidePaddingPercent: 0.18,
        arcBottomExtent: 0.72,
    },
    _a[GaugeType.Grafana] = {
        topPaddingPercent: 0.18,
        bottomPaddingPercent: 0.08,
        sidePaddingPercent: 0.16,
        arcBottomExtent: 0.70,
    },
    _a);
/**
 * Calculates optimal outer radius given available space and gauge type
 * This is the core function that determines how much space the gauge will use
 *
 * CRITICAL: The resulting viewBox must fit within parentWidth x parentHeight
 */
export var calculateOptimalRadius = function (parentWidth, parentHeight, gaugeType, marginPercent) {
    if (marginPercent === void 0) { marginPercent = 0; }
    var config = GAUGE_TYPE_CONFIG[gaugeType];
    // Apply margin (as percentage of dimensions)
    var availableWidth = parentWidth * (1 - marginPercent);
    var availableHeight = parentHeight * (1 - marginPercent);
    var widthMultiplier;
    var heightMultiplier;
    // All gauge types now use separate padding values
    var typedConfig = config;
    // width = diameter + 2*sidePadding
    widthMultiplier = 2 + 2 * typedConfig.sidePaddingPercent;
    if (gaugeType === GaugeType.Semicircle) {
        // Semicircle: topPadding + radius (to center) + bottomPadding
        heightMultiplier = typedConfig.topPaddingPercent + 1 + typedConfig.bottomPaddingPercent;
    }
    else {
        // Radial/Grafana: topPadding + radius + arcBottomExtent*radius + bottomPadding
        var arcBottom = typedConfig.arcBottomExtent || 0.72;
        heightMultiplier = typedConfig.topPaddingPercent + 1 + arcBottom + typedConfig.bottomPaddingPercent;
    }
    var radiusFromWidth = availableWidth / widthMultiplier;
    var radiusFromHeight = availableHeight / heightMultiplier;
    // Use the smaller radius to ensure gauge fits in both dimensions
    return Math.min(radiusFromWidth, radiusFromHeight);
};
/**
 * Calculates the viewBox dimensions based on gauge type and radius
 * ViewBox defines the coordinate system for the SVG
 */
export var calculateViewBox = function (outerRadius, gaugeType) {
    var config = GAUGE_TYPE_CONFIG[gaugeType];
    var diameter = outerRadius * 2;
    // All gauge types now use separate padding values
    var typedConfig = config;
    var topPadding = outerRadius * typedConfig.topPaddingPercent;
    var bottomPadding = outerRadius * typedConfig.bottomPaddingPercent;
    var sidePadding = outerRadius * typedConfig.sidePaddingPercent;
    var width = diameter + sidePadding * 2;
    var height;
    if (gaugeType === GaugeType.Semicircle) {
        // Semicircle: topPadding + radius (to center) + bottomPadding
        height = topPadding + outerRadius + bottomPadding;
    }
    else {
        // Radial/Grafana: topPadding + radius + arcBottomExtent*radius + bottomPadding
        var arcBottom = typedConfig.arcBottomExtent || 0.72;
        height = topPadding + outerRadius + (outerRadius * arcBottom) + bottomPadding;
    }
    return {
        x: 0,
        y: 0,
        width: width,
        height: height,
        toString: function () {
            return "".concat(this.x, " ").concat(this.y, " ").concat(this.width, " ").concat(this.height);
        },
    };
};
/**
 * Calculates the center point of the gauge within the viewBox
 * This is where the doughnut element should be positioned
 */
export var calculateGaugeCenter = function (viewBox, outerRadius, gaugeType) {
    var config = GAUGE_TYPE_CONFIG[gaugeType];
    // All gauge types now use separate padding values
    var typedConfig = config;
    // Horizontal center is always in the middle
    var x = viewBox.width / 2;
    // Vertical center: topPadding + radius from the top
    var topPadding = outerRadius * typedConfig.topPaddingPercent;
    var y = topPadding + outerRadius;
    return { x: x, y: y };
};
/**
 * Calculates complete layout information for the gauge
 * This is the main function used by the rendering code
 */
export var calculateGaugeLayout = function (parentWidth, parentHeight, gaugeType, arcWidth, marginPercent) {
    if (marginPercent === void 0) { marginPercent = 0; }
    // Calculate optimal outer radius
    var outerRadius = calculateOptimalRadius(parentWidth, parentHeight, gaugeType, marginPercent);
    // Calculate inner radius based on arc width
    var innerRadius = outerRadius * (1 - arcWidth);
    // Calculate viewBox
    var viewBox = calculateViewBox(outerRadius, gaugeType);
    // Calculate gauge center position
    var gaugeCenter = calculateGaugeCenter(viewBox, outerRadius, gaugeType);
    // The doughnut transform is the same as gauge center
    // because doughnut is positioned relative to its parent
    var doughnutTransform = { x: outerRadius, y: outerRadius };
    return {
        viewBox: viewBox,
        outerRadius: outerRadius,
        innerRadius: innerRadius,
        gaugeCenter: gaugeCenter,
        doughnutTransform: doughnutTransform,
    };
};
/**
 * NEW: Calculates optimized layout using dynamic padding based on actual config
 * This maximizes gauge size by only reserving space that's actually needed
 */
export var calculateOptimizedLayout = function (parentWidth, parentHeight, gaugeType, arcWidth, paddingConfig, marginPercent) {
    if (marginPercent === void 0) { marginPercent = 0; }
    // Apply margin
    var availableWidth = parentWidth * (1 - marginPercent);
    var availableHeight = parentHeight * (1 - marginPercent);
    // Arc bottom extent for radial/grafana types
    var arcBottomExtent = gaugeType === GaugeType.Semicircle ? 0 :
        (gaugeType === GaugeType.Grafana ? 0.70 : 0.72);
    // First pass: estimate radius to calculate dynamic padding
    // Use a conservative estimate (50% of min dimension)
    var estimatedRadius = Math.min(availableWidth, availableHeight) * 0.4;
    var padding = calculateDynamicPadding(gaugeType, paddingConfig, estimatedRadius);
    // Calculate radius that fits with the dynamic padding
    // width = diameter + 2*sidePadding => radius = (width - 2*sidePadding) / 2
    // height = topPadding + radius + (arcBottomExtent * radius) + bottomPadding
    //        = topPadding + radius * (1 + arcBottomExtent) + bottomPadding
    // => radius = (height - topPadding - bottomPadding) / (1 + arcBottomExtent)
    var radiusFromWidth = (availableWidth - 2 * padding.side) / 2;
    var radiusFromHeight;
    if (gaugeType === GaugeType.Semicircle) {
        // Semicircle: height = topPadding + radius + bottomPadding
        radiusFromHeight = availableHeight - padding.top - padding.bottom;
    }
    else {
        // Radial/Grafana: height = topPadding + radius * (1 + arcBottomExtent) + bottomPadding
        radiusFromHeight = (availableHeight - padding.top - padding.bottom) / (1 + arcBottomExtent);
    }
    var outerRadius = Math.max(10, Math.min(radiusFromWidth, radiusFromHeight));
    var innerRadius = outerRadius * (1 - arcWidth);
    // Recalculate padding with actual radius for precision
    var finalPadding = calculateDynamicPadding(gaugeType, paddingConfig, outerRadius);
    // Calculate viewBox with actual padding
    var diameter = outerRadius * 2;
    var viewBoxWidth = diameter + finalPadding.side * 2;
    var viewBoxHeight;
    if (gaugeType === GaugeType.Semicircle) {
        viewBoxHeight = finalPadding.top + outerRadius + finalPadding.bottom;
    }
    else {
        viewBoxHeight = finalPadding.top + outerRadius + (outerRadius * arcBottomExtent) + finalPadding.bottom;
    }
    var viewBox = {
        x: 0,
        y: 0,
        width: viewBoxWidth,
        height: viewBoxHeight,
        toString: function () {
            return "".concat(this.x, " ").concat(this.y, " ").concat(this.width, " ").concat(this.height);
        },
    };
    // Calculate gauge center
    var gaugeCenter = {
        x: viewBoxWidth / 2,
        y: finalPadding.top + outerRadius,
    };
    return {
        viewBox: viewBox,
        outerRadius: outerRadius,
        innerRadius: innerRadius,
        gaugeCenter: gaugeCenter,
        doughnutTransform: { x: outerRadius, y: outerRadius },
    };
};
/**
 * Updates the dimensions object with new layout calculations
 * This maintains backward compatibility with existing code
 */
export var updateDimensionsFromLayout = function (dimensions, layout) {
    dimensions.outerRadius = layout.outerRadius;
    dimensions.innerRadius = layout.innerRadius;
    dimensions.width = layout.viewBox.width;
    dimensions.height = layout.viewBox.height;
    // Keep margin values as they might be used elsewhere
    // But they should be derived from the layout in the future
};
/**
 * Validates that a gauge won't cause infinite resizing
 * Returns true if the layout is stable
 *
 * Compares radius, viewBox dimensions, and center position to detect oscillation
 */
export var isLayoutStable = function (previousLayout, currentLayout, tolerance) {
    if (tolerance === void 0) { tolerance = 0.01; }
    if (!previousLayout)
        return true;
    // Check radius stability
    var radiusDiff = Math.abs(currentLayout.outerRadius - previousLayout.outerRadius);
    var radiusChange = previousLayout.outerRadius > 0
        ? radiusDiff / previousLayout.outerRadius
        : radiusDiff;
    // Check viewBox stability
    var viewBoxWidthDiff = Math.abs(currentLayout.viewBox.width - previousLayout.viewBox.width);
    var viewBoxHeightDiff = Math.abs(currentLayout.viewBox.height - previousLayout.viewBox.height);
    var viewBoxWidthChange = previousLayout.viewBox.width > 0
        ? viewBoxWidthDiff / previousLayout.viewBox.width
        : viewBoxWidthDiff;
    var viewBoxHeightChange = previousLayout.viewBox.height > 0
        ? viewBoxHeightDiff / previousLayout.viewBox.height
        : viewBoxHeightDiff;
    // Check center position stability
    var centerXDiff = Math.abs(currentLayout.gaugeCenter.x - previousLayout.gaugeCenter.x);
    var centerYDiff = Math.abs(currentLayout.gaugeCenter.y - previousLayout.gaugeCenter.y);
    var centerXChange = previousLayout.gaugeCenter.x > 0
        ? centerXDiff / previousLayout.gaugeCenter.x
        : centerXDiff;
    var centerYChange = previousLayout.gaugeCenter.y > 0
        ? centerYDiff / previousLayout.gaugeCenter.y
        : centerYDiff;
    // All properties must be within tolerance for layout to be stable
    return radiusChange < tolerance &&
        viewBoxWidthChange < tolerance &&
        viewBoxHeightChange < tolerance &&
        centerXChange < tolerance &&
        centerYChange < tolerance;
};
/**
 * TWO-PASS RENDERING: Calculate optimized layout from measured bounds
 *
 * After the first render, we measure the actual bounding box of the <g> element.
 * This tells us exactly how much space the gauge content uses.
 * We then recalculate the viewBox to fit the content with minimal padding.
 *
 * The measured bounds are relative to the <g> element's transform origin (gauge center).
 * So bounds.x and bounds.y can be negative (content extends left/up from center).
 */
export var calculateLayoutFromMeasuredBounds = function (parentWidth, parentHeight, measuredBounds, gaugeType, arcWidth, currentLayout) {
    // Add safety margin around the measured bounds
    // For small containers, use smaller margins to maximize gauge size
    var minDimension = Math.min(parentWidth, parentHeight);
    // Scale margin factor: 3% for small containers (<150px), up to 5% for larger ones
    var marginFactor = minDimension < 150 ? 0.02 :
        minDimension < 250 ? 0.03 : 0.05;
    // Scale minimum margin with container size: 2-5px
    var minMargin = Math.max(2, Math.min(5, minDimension * 0.02));
    var marginX = Math.max(minMargin, measuredBounds.width * marginFactor);
    var marginY = Math.max(minMargin, measuredBounds.height * marginFactor);
    // The measured bounds are relative to the gauge center (0,0 in <g> space)
    // bounds.x is the left edge, bounds.y is the top edge (can be negative)
    // We need to calculate the full extent from center
    var leftExtent = Math.abs(measuredBounds.x) + marginX;
    var rightExtent = measuredBounds.x + measuredBounds.width + marginX;
    var topExtent = Math.abs(measuredBounds.y) + marginY;
    var bottomExtent = measuredBounds.y + measuredBounds.height + marginY;
    // Use SYMMETRIC side padding to keep gauge visually centered
    // This handles asymmetric content (like a tick on only one side) by 
    // using the maximum extent on both sides
    var maxSideExtent = Math.max(leftExtent, rightExtent);
    var contentWidth = maxSideExtent * 2; // Symmetric width
    var contentHeight = topExtent + bottomExtent;
    // Calculate the scale factor to fit the content in the parent
    var scaleX = parentWidth / contentWidth;
    var scaleY = parentHeight / contentHeight;
    var scale = Math.min(scaleX, scaleY);
    // Calculate new outer radius based on the scale
    var newOuterRadius = currentLayout.outerRadius * scale;
    var newInnerRadius = newOuterRadius * (1 - arcWidth);
    // Calculate new viewBox dimensions
    var newViewBoxWidth = contentWidth * scale;
    var newViewBoxHeight = contentHeight * scale;
    // Calculate new center position
    // With symmetric side padding, center is always in the horizontal middle
    var newCenterX = newViewBoxWidth / 2;
    var newCenterY = topExtent * scale;
    var viewBox = {
        x: 0,
        y: 0,
        width: newViewBoxWidth,
        height: newViewBoxHeight,
        toString: function () {
            return "".concat(this.x, " ").concat(this.y, " ").concat(this.width, " ").concat(this.height);
        },
    };
    return {
        viewBox: viewBox,
        outerRadius: newOuterRadius,
        innerRadius: newInnerRadius,
        gaugeCenter: { x: newCenterX, y: newCenterY },
        doughnutTransform: { x: newOuterRadius, y: newOuterRadius },
    };
};
/**
 * TWO-PASS RENDERING: Optimized approach using actual measured bounds
 *
 * Pass 1: Render with generous initial padding to ensure all content is visible
 *         for accurate measurement. This pass is hidden from the user.
 * Pass 2: Recalculate viewBox to maximize gauge size while fitting content
 *
 * This approach:
 * 1. Uses generous padding for first pass (hidden) to capture all content
 * 2. Renders all elements (arcs, labels, pointer)
 * 3. Measures the actual bounding box of the rendered content
 * 4. Adjusts the viewBox to tightly fit the content with proper margins
 */
export var calculateTightLayout = function (parentWidth, parentHeight, gaugeType, arcWidth, marginPercent) {
    if (marginPercent === void 0) { marginPercent = 0; }
    // Apply margin
    var availableWidth = parentWidth * (1 - marginPercent);
    var availableHeight = parentHeight * (1 - marginPercent);
    // Arc bottom extent for radial/grafana types
    // Semicircle needs extra bottom space for needle/value label
    var arcBottomExtent = gaugeType === GaugeType.Semicircle ? 0.15 :
        (gaugeType === GaugeType.Grafana ? 0.70 : 0.72);
    // Use PROPORTIONAL padding for first pass
    // This ensures gauges stay usable at all sizes
    // For small containers (<200px), use 8% padding; for larger, cap at reasonable pixel values
    var minDimension = Math.min(availableWidth, availableHeight);
    // Proportional padding: 10% of min dimension, capped between 8-40px
    var topPadding = Math.max(8, Math.min(30, minDimension * 0.10));
    var sidePadding = Math.max(10, Math.min(40, minDimension * 0.12));
    var bottomPadding = Math.max(6, Math.min(25, minDimension * 0.08));
    // Calculate radius with proportional padding
    var radiusFromWidth = (availableWidth - 2 * sidePadding) / 2;
    var radiusFromHeight;
    if (gaugeType === GaugeType.Semicircle) {
        // Semicircle: top padding + radius + bottom padding (for value label/needle)
        radiusFromHeight = (availableHeight - topPadding - bottomPadding);
    }
    else {
        // Radial/Grafana: top padding + radius + arc bottom extent + bottom padding
        radiusFromHeight = (availableHeight - topPadding - bottomPadding) / (1 + arcBottomExtent);
    }
    var outerRadius = Math.max(10, Math.min(radiusFromWidth, radiusFromHeight));
    var innerRadius = outerRadius * (1 - arcWidth);
    // Calculate viewBox with generous padding
    var diameter = outerRadius * 2;
    var viewBoxWidth = diameter + sidePadding * 2;
    var viewBoxHeight;
    if (gaugeType === GaugeType.Semicircle) {
        // Semicircle: top padding + radius + bottom padding
        viewBoxHeight = topPadding + outerRadius + bottomPadding;
    }
    else {
        // Radial/Grafana: top padding + radius + arc extent + bottom padding
        viewBoxHeight = topPadding + outerRadius + (outerRadius * arcBottomExtent) + bottomPadding;
    }
    var viewBox = {
        x: 0,
        y: 0,
        width: viewBoxWidth,
        height: viewBoxHeight,
        toString: function () {
            return "".concat(this.x, " ").concat(this.y, " ").concat(this.width, " ").concat(this.height);
        },
    };
    // Calculate gauge center
    var gaugeCenter = {
        x: viewBoxWidth / 2,
        y: topPadding + outerRadius,
    };
    return {
        viewBox: viewBox,
        outerRadius: outerRadius,
        innerRadius: innerRadius,
        gaugeCenter: gaugeCenter,
        doughnutTransform: { x: outerRadius, y: outerRadius },
    };
};
