// react-gauge-component/dist/lib/GaugeComponent/hooks/labels.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as utils from './utils';
import CONSTANTS from '../constants';
import { defaultTickLabels } from '../types/Tick';
import { line } from 'd3-shape';
import { GaugeType } from '../types/GaugeComponentProps';
import { getArcDataByValue, getCoordByValue, getEffectiveAngles } from './arc';
var registerCustomContentItem = function (gauge, domNode, renderContent, value, arcColor) {
    if (!gauge.customContent) {
        gauge.customContent = { current: {} };
    }
    var current = gauge.customContent.current;
    if (!Array.isArray(current.items)) {
        current.items = [];
    }
    current.items.push({ domNode: domNode, renderContent: renderContent, value: value, arcColor: arcColor });
    // Backward compatibility: keep single-node fields in sync with the most recent item
    current.domNode = domNode;
    current.renderContent = renderContent;
    current.value = value;
    current.arcColor = arcColor;
};
export var setupLabels = function (gauge) {
    var _a;
    // Reset custom content items for this render pass (value label + tick labels)
    if ((_a = gauge.customContent) === null || _a === void 0 ? void 0 : _a.current) {
        var cc = gauge.customContent.current;
        cc.items = [];
    }
    setupValueLabel(gauge);
    setupTicks(gauge);
};
export var setupValueLabel = function (gauge) {
    var _a;
    var labels = gauge.props.labels;
    if (!((_a = labels === null || labels === void 0 ? void 0 : labels.valueLabel) === null || _a === void 0 ? void 0 : _a.hide))
        addValueText(gauge);
};
export var setupTicks = function (gauge) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    var labels = gauge.props.labels;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    // Pre-compute radial offsets for auto-spacing closely-spaced ticks
    var autoSpaceOffsets = computeAutoSpaceOffsets(gauge);
    var tickRenderIndex = 0;
    if (CONSTANTS.debugTicksRadius) {
        for (var index = 0; index < maxValue; index++) {
            var indexTick = mapTick(index, gauge);
            addTick(indexTick, gauge, tickRenderIndex++);
        }
    }
    else if (!((_a = labels.tickLabels) === null || _a === void 0 ? void 0 : _a.hideMinMax)) {
        var alreadyHaveMinValueTick = (_c = (_b = labels.tickLabels) === null || _b === void 0 ? void 0 : _b.ticks) === null || _c === void 0 ? void 0 : _c.some(function (tick) { return tick.value == minValue; });
        if (!alreadyHaveMinValueTick) {
            //Add min value tick
            var minValueTick = mapTick(minValue, gauge);
            addTick(minValueTick, gauge, tickRenderIndex++);
        }
        var alreadyHaveMaxValueTick = (_e = (_d = labels.tickLabels) === null || _d === void 0 ? void 0 : _d.ticks) === null || _e === void 0 ? void 0 : _e.some(function (tick) { return tick.value == maxValue; });
        if (!alreadyHaveMaxValueTick) {
            // //Add max value tick
            var maxValueTick = mapTick(maxValue, gauge);
            addTick(maxValueTick, gauge, tickRenderIndex++);
        }
    }
    if (((_g = (_f = labels.tickLabels) === null || _f === void 0 ? void 0 : _f.ticks) === null || _g === void 0 ? void 0 : _g.length) > 0) {
        (_j = (_h = labels.tickLabels) === null || _h === void 0 ? void 0 : _h.ticks) === null || _j === void 0 ? void 0 : _j.forEach(function (tick) {
            // Attach auto-space offset if computed
            var tickValue = tick.value;
            if (autoSpaceOffsets.has(tickValue)) {
                tick.__autoSpaceOffset = autoSpaceOffsets.get(tickValue);
            }
            addTick(tick, gauge, tickRenderIndex++);
        });
    }
    addArcTicks(gauge, function () { return tickRenderIndex++; });
};
export var addArcTicks = function (gauge, nextIndex) {
    var _a;
    (_a = gauge.arcData.current) === null || _a === void 0 ? void 0 : _a.map(function (subArc) {
        if (subArc.showTick)
            return subArc.limit;
    }).forEach(function (tickValue) {
        if (tickValue !== undefined && tickValue !== null)
            addTick(mapTick(tickValue, gauge), gauge, nextIndex ? nextIndex() : undefined);
    });
};
/**
 * Compute radial offsets for closely-spaced tick labels to prevent overlap.
 * Returns a Map of tickValue -> offset direction (+1 = outward, -1 = inward).
 */
var computeAutoSpaceOffsets = function (gauge) {
    var _a, _b, _c, _d, _e;
    var offsets = new Map();
    var labels = gauge.props.labels;
    // Only compute if autoSpaceTickLabels is enabled
    if (!((_a = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _a === void 0 ? void 0 : _a.autoSpaceTickLabels)) {
        return offsets;
    }
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var range = maxValue - minValue;
    if (range <= 0)
        return offsets;
    // Collect all tick values (from ticks array + min/max if not hidden)
    var tickValues = [];
    if (!((_b = labels.tickLabels) === null || _b === void 0 ? void 0 : _b.hideMinMax)) {
        var hasTicks = ((_c = labels.tickLabels) === null || _c === void 0 ? void 0 : _c.ticks) || [];
        if (!hasTicks.some(function (t) { return t.value === minValue; }))
            tickValues.push(minValue);
        if (!hasTicks.some(function (t) { return t.value === maxValue; }))
            tickValues.push(maxValue);
    }
    (_e = (_d = labels.tickLabels) === null || _d === void 0 ? void 0 : _d.ticks) === null || _e === void 0 ? void 0 : _e.forEach(function (tick) {
        if (tick.value !== undefined)
            tickValues.push(tick.value);
    });
    // Sort by value
    tickValues.sort(function (a, b) { return a - b; });
    // Threshold: if two ticks are closer than 3% of range, they need spacing
    var proximityThreshold = range * 0.03;
    // Detect clusters of close ticks and assign alternating offsets
    for (var i = 0; i < tickValues.length - 1; i++) {
        var current = tickValues[i];
        var next = tickValues[i + 1];
        var distance = next - current;
        if (distance < proximityThreshold) {
            // These two are too close - assign angular offsets to shift them along the arc
            // First tick: shift slightly counter-clockwise (-1)
            // Second tick: shift slightly clockwise (+1)
            if (!offsets.has(current)) {
                offsets.set(current, -1);
            }
            if (!offsets.has(next)) {
                offsets.set(next, 1);
            }
        }
    }
    return offsets;
};
export var mapTick = function (value, gauge) {
    var tickLabels = gauge.props.labels.tickLabels;
    return {
        value: value,
        valueConfig: tickLabels === null || tickLabels === void 0 ? void 0 : tickLabels.defaultTickValueConfig,
        lineConfig: tickLabels === null || tickLabels === void 0 ? void 0 : tickLabels.defaultTickLineConfig
    };
};
/**
 * Calculate scale factor based on gauge radius to make ticks proportional
 * Reference radius is 100px (default behavior preserved at this size)
 */
var getTickScaleFactor = function (gauge) {
    var referenceRadius = 100;
    var outerRadius = gauge.dimensions.current.outerRadius;
    // Scale proportionally but clamp between 0.5 and 1.5 to avoid extremes
    return Math.max(0.5, Math.min(1.5, outerRadius / referenceRadius));
};
export var addTickLine = function (tick, gauge) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
    var labels = gauge.props.labels;
    var _u = calculateAnchorAndAngleByValue(tick === null || tick === void 0 ? void 0 : tick.value, gauge), tickAnchor = _u.tickAnchor, angle = _u.angle;
    // Get scale factor for proportional sizing
    var scaleFactor = getTickScaleFactor(gauge);
    var tickDistanceFromArc = ((_a = tick.lineConfig) === null || _a === void 0 ? void 0 : _a.distanceFromArc) || ((_c = (_b = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _b === void 0 ? void 0 : _b.defaultTickLineConfig) === null || _c === void 0 ? void 0 : _c.distanceFromArc) || 0;
    // Scale distance from arc
    tickDistanceFromArc = tickDistanceFromArc * scaleFactor;
    // For outer ticks: tick starts at outer arc edge + distance, line goes outward
    // For inner ticks: tick starts at inner arc edge - distance, line goes inward toward center
    // Both use negative offset to push START position away from center (toward arc edge)
    if (((_e = (_d = gauge.props.labels) === null || _d === void 0 ? void 0 : _d.tickLabels) === null || _e === void 0 ? void 0 : _e.type) == "outer") {
        tickDistanceFromArc = -tickDistanceFromArc;
    }
    else {
        // Inner ticks: start at arc edge (small gap), not deep inside
        tickDistanceFromArc = tickDistanceFromArc; // Keep positive to offset slightly from arc
    }
    var coords = getLabelCoordsByValue(tick === null || tick === void 0 ? void 0 : tick.value, gauge, tickDistanceFromArc);
    var tickColor = ((_f = tick.lineConfig) === null || _f === void 0 ? void 0 : _f.color) || ((_h = (_g = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _g === void 0 ? void 0 : _g.defaultTickLineConfig) === null || _h === void 0 ? void 0 : _h.color) || ((_j = defaultTickLabels.defaultTickLineConfig) === null || _j === void 0 ? void 0 : _j.color);
    var tickWidth = ((_k = tick.lineConfig) === null || _k === void 0 ? void 0 : _k.width) || ((_m = (_l = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _l === void 0 ? void 0 : _l.defaultTickLineConfig) === null || _m === void 0 ? void 0 : _m.width) || ((_o = defaultTickLabels.defaultTickLineConfig) === null || _o === void 0 ? void 0 : _o.width);
    var tickLength = ((_p = tick.lineConfig) === null || _p === void 0 ? void 0 : _p.length) || ((_r = (_q = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _q === void 0 ? void 0 : _q.defaultTickLineConfig) === null || _r === void 0 ? void 0 : _r.length) || ((_s = defaultTickLabels.defaultTickLineConfig) === null || _s === void 0 ? void 0 : _s.length);
    // Scale tick dimensions proportionally with gauge size
    tickLength = tickLength * scaleFactor;
    tickWidth = Math.max(0.5, tickWidth * scaleFactor);
    // Calculate the end coordinates based on tick type
    // coords is positioned at the arc edge
    // angle points OUTWARD from center (away from arc center)
    var endX;
    var endY;
    var angleRad = (angle * Math.PI) / 180;
    // For outer ticks: line goes OUTWARD from arc (in direction of angle)
    // For inner ticks: line goes INWARD toward center (opposite of angle)
    if (((_t = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _t === void 0 ? void 0 : _t.type) == "inner") {
        // Inner ticks: start at inner arc edge, draw toward center (opposite direction)
        endX = coords.x - tickLength * Math.cos(angleRad);
        endY = coords.y - tickLength * Math.sin(angleRad);
    }
    else {
        // Outer ticks: start at outer arc edge, draw away from center (same direction as angle)
        endX = coords.x + tickLength * Math.cos(angleRad);
        endY = coords.y + tickLength * Math.sin(angleRad);
    }
    // (gauge.dimensions.current.outerRadius - gauge.dimensions.current.innerRadius)
    // Create a D3 line generator
    var lineGenerator = line();
    var lineCoordinates;
    // Define the line coordinates
    lineCoordinates = [[coords.x, coords.y], [endX, endY]];
    // Append a path element for the line
    gauge.g.current
        .append("path")
        .datum(lineCoordinates)
        .attr("class", CONSTANTS.tickLineClassname)
        .attr("d", lineGenerator)
        // .attr("transform", `translate(${0}, ${0})`)
        .attr("stroke", tickColor)
        .attr("stroke-width", tickWidth)
        .attr("fill", "none");
    // .attr("stroke-linecap", "round")
    // .attr("stroke-linejoin", "round")
    // .attr("transform", `rotate(${angle})`);
};
export var addTickValue = function (tick, gauge) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
    var labels = gauge.props.labels;
    var tickValue = tick === null || tick === void 0 ? void 0 : tick.value;
    var _t = calculateAnchorAndAngleByValue(tickValue, gauge), tickAnchor = _t.tickAnchor, angle = _t.angle;
    // Get scale factor for proportional sizing
    var scaleFactor = getTickScaleFactor(gauge);
    var isInner = ((_a = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _a === void 0 ? void 0 : _a.type) == "inner";
    var tickDistanceFromArc = ((_b = tick.lineConfig) === null || _b === void 0 ? void 0 : _b.distanceFromArc) || ((_d = (_c = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _c === void 0 ? void 0 : _c.defaultTickLineConfig) === null || _d === void 0 ? void 0 : _d.distanceFromArc) || 0;
    var tickLength = ((_e = tick.lineConfig) === null || _e === void 0 ? void 0 : _e.length) || ((_g = (_f = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _f === void 0 ? void 0 : _f.defaultTickLineConfig) === null || _g === void 0 ? void 0 : _g.length) || 0;
    // Scale distance and length
    tickDistanceFromArc = tickDistanceFromArc * scaleFactor;
    tickLength = tickLength * scaleFactor;
    // Calculate label position: at the end of the tick line + padding for text
    var _shouldHideTickLine = shouldHideTickLine(tick, gauge);
    // For outer ticks, centerToArcLengthSubtract is negative to push outward
    // For inner ticks, it's positive to push inward
    var tickLineOffset = _shouldHideTickLine ? 0 : (tickDistanceFromArc + tickLength);
    // Get distance from text prop - check for explicit user config first
    // The default config has distanceFromText: 2 which is fine for outer but too small for inner
    var userDistanceFromText = (_j = (_h = tick.lineConfig) === null || _h === void 0 ? void 0 : _h.distanceFromText) !== null && _j !== void 0 ? _j : (_l = (_k = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _k === void 0 ? void 0 : _k.defaultTickLineConfig) === null || _l === void 0 ? void 0 : _l.distanceFromText;
    // Inner ticks need significantly more padding because:
    // 1. Text is inside the arc where space is limited
    // 2. Tick lines point toward center, so text must be pushed further inward
    // 3. Overlapping is more visually noticeable in the confined inner space
    // Only use the small default (2) for outer ticks; use larger (35) for inner
    var minInnerPadding = 12;
    var textPadding;
    if (isInner) {
        // For inner: use max of user config and minimum inner padding to prevent overlap
        textPadding = Math.max(userDistanceFromText !== null && userDistanceFromText !== void 0 ? userDistanceFromText : minInnerPadding, minInnerPadding) * scaleFactor;
    }
    else {
        // For outer: use user config or small default
        textPadding = (userDistanceFromText !== null && userDistanceFromText !== void 0 ? userDistanceFromText : 5) * scaleFactor;
    }
    var centerToArcLengthSubtract;
    if (isInner) {
        // Inner: push position inward (toward center) past tick line
        centerToArcLengthSubtract = tickLineOffset + textPadding;
    }
    else {
        // Outer: push position outward (away from center) past tick line
        centerToArcLengthSubtract = -(tickLineOffset + textPadding);
    }
    var coords = getLabelCoordsByValue(tickValue, gauge, centerToArcLengthSubtract);
    // Apply auto-space angular offset for closely-spaced ticks
    // This shifts the label position along the arc curve (not radially)
    var autoSpaceOffset = tick.__autoSpaceOffset;
    if (typeof autoSpaceOffset === 'number') {
        // Angular offset in degrees - shift along the arc
        var angularOffsetDegrees = 4 * autoSpaceOffset; // ~4 degrees per direction
        var angleRad = (angle * Math.PI) / 180;
        var offsetAngleRad = ((angle + angularOffsetDegrees) * Math.PI) / 180;
        // Calculate tangent direction and apply offset along the arc
        var tangentX = -Math.sin(angleRad);
        var tangentY = Math.cos(angleRad);
        var offsetAmount = angularOffsetDegrees * 0.5 * scaleFactor; // Scale with gauge size
        coords.x += tangentX * offsetAmount;
        coords.y += tangentY * offsetAmount;
    }
    var defaultTickValueConfig = ((_m = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _m === void 0 ? void 0 : _m.defaultTickValueConfig) || defaultTickLabels.defaultTickValueConfig;
    var tickValueConfig = tick.valueConfig || defaultTickValueConfig;
    // IMPORTANT: Merge styles so per-tick style overrides don't wipe defaults (e.g. empty {} -> black text)
    var defaultStyle = (defaultTickValueConfig === null || defaultTickValueConfig === void 0 ? void 0 : defaultTickValueConfig.style) || {};
    var tickOverrideStyle = ((_o = tick.valueConfig) === null || _o === void 0 ? void 0 : _o.style) || {};
    var tickValueStyle = __assign(__assign({}, defaultStyle), tickOverrideStyle);
    // If configured, allow tick labels to be rendered as custom React content
    var arcColor = ((_p = getArcDataByValue(tickValue, gauge)) === null || _p === void 0 ? void 0 : _p.color) || "white";
    if (tickValueConfig === null || tickValueConfig === void 0 ? void 0 : tickValueConfig.renderContent) {
        addCustomTickContent(gauge, tickValueConfig, tickValue, arcColor, coords.x, coords.y);
        return;
    }
    var text = '';
    var maxDecimalDigits = ((_q = tick.valueConfig) === null || _q === void 0 ? void 0 : _q.maxDecimalDigits) || (defaultTickValueConfig === null || defaultTickValueConfig === void 0 ? void 0 : defaultTickValueConfig.maxDecimalDigits);
    if ((_r = tick.valueConfig) === null || _r === void 0 ? void 0 : _r.formatTextValue) {
        text = tick.valueConfig.formatTextValue(utils.floatingNumber(tickValue, maxDecimalDigits));
    }
    else if (defaultTickValueConfig === null || defaultTickValueConfig === void 0 ? void 0 : defaultTickValueConfig.formatTextValue) {
        text = defaultTickValueConfig.formatTextValue(utils.floatingNumber(tickValue, maxDecimalDigits));
    }
    else if (gauge.props.minValue === 0 && gauge.props.maxValue === 100) {
        text = utils.floatingNumber(tickValue, maxDecimalDigits).toString();
        text += "%";
    }
    else {
        text = utils.floatingNumber(tickValue, maxDecimalDigits).toString();
    }
    // Scale position offsets
    var positionOffset = 10 * scaleFactor;
    if (((_s = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _s === void 0 ? void 0 : _s.type) == "inner") {
        if (tickAnchor === "end")
            coords.x += positionOffset;
        if (tickAnchor === "start")
            coords.x -= positionOffset;
    }
    else {
        if (tickAnchor === "middle")
            coords.y += 2 * scaleFactor;
    }
    if (tickAnchor === "middle") {
        coords.y += 0;
    }
    else {
        coords.y += 3 * scaleFactor;
    }
    // Scale font size based on gauge size
    var baseFontSize = parseFloat(String(tickValueStyle.fontSize || '12px')) || 12;
    var scaledFontSize = Math.max(6, baseFontSize * scaleFactor);
    tickValueStyle.fontSize = "".concat(scaledFontSize, "px");
    tickValueStyle.textAnchor = tickAnchor;
    addText(text, coords.x, coords.y, gauge, tickValueStyle, CONSTANTS.tickValueClassname);
};
export var addTick = function (tick, gauge, renderIndex) {
    var labels = gauge.props.labels;
    // Attach render index for staggering calculations without changing public API
    tick.__renderIndex = renderIndex;
    //Make validation for sequence of values respecting DEFAULT -> DEFAULT FROM USER -> SPECIFIC TICK VALUE
    var _shouldHideTickLine = shouldHideTickLine(tick, gauge);
    var _shouldHideTickValue = shouldHideTickValue(tick, gauge);
    if (!_shouldHideTickLine)
        addTickLine(tick, gauge);
    if (!CONSTANTS.debugTicksRadius && !_shouldHideTickValue) {
        addTickValue(tick, gauge);
    }
};
export var getLabelCoordsByValue = function (value, gauge, centerToArcLengthSubtract) {
    var _a;
    if (centerToArcLengthSubtract === void 0) { centerToArcLengthSubtract = 0; }
    var labels = gauge.props.labels;
    var type = (_a = labels.tickLabels) === null || _a === void 0 ? void 0 : _a.type;
    var _b = getCoordByValue(value, gauge, type, centerToArcLengthSubtract), x = _b.x, y = _b.y;
    return { x: x, y: y };
};
export var addText = function (html, x, y, gauge, style, className, rotate) {
    if (rotate === void 0) { rotate = 0; }
    var div = gauge.g.current
        .append("g")
        .attr("class", className)
        .attr("transform", "translate(".concat(x, ", ").concat(y, ")"))
        .append("text")
        .text(html); // use html() instead of text()
    applyTextStyles(div, style);
    div.attr("transform", "rotate(".concat(rotate, ")"));
};
var applyTextStyles = function (div, style) {
    //Apply default styles
    Object.entries(style).forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return div.style(utils.camelCaseToKebabCase(key), value);
    });
    //Apply custom styles
    if (style != undefined)
        Object.entries(style).forEach(function (_a) {
            var key = _a[0], value = _a[1];
            return div.style(utils.camelCaseToKebabCase(key), value);
        });
};
//Adds text undeneath the graft to display which percentage is the current one
export var addValueText = function (gauge) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var _l = gauge.props, labels = _l.labels, pointers = _l.pointers;
    var valueLabel = ((labels === null || labels === void 0 ? void 0 : labels.valueLabel) || {});
    // Check for multi-pointer mode
    var isMultiPointer = Array.isArray(pointers) && pointers.length > 0;
    var multiPointerDisplay = (_a = valueLabel.multiPointerDisplay) !== null && _a !== void 0 ? _a : 'primary';
    if (isMultiPointer && multiPointerDisplay === 'none') {
        return; // Hide value label in multi-pointer mode if configured
    }
    // Position label in the CENTER of the gauge arc
    var innerRadius = gauge.dimensions.current.innerRadius;
    var x = 0;
    var y = 0;
    // Default positioning - moved lower to avoid overlapping with needle
    if (gauge.props.type == GaugeType.Semicircle) {
        y = innerRadius * -0.15;
    }
    else if (gauge.props.type == GaugeType.Radial) {
        y = innerRadius * 0.4;
    }
    else if (gauge.props.type == GaugeType.Grafana) {
        y = innerRadius * 0.35;
    }
    // Apply user offsets
    x += (_b = valueLabel.offsetX) !== null && _b !== void 0 ? _b : 0;
    y += (_c = valueLabel.offsetY) !== null && _c !== void 0 ? _c : 0;
    // Check if animateValue is enabled and animation will occur
    // If so, show the starting value (prevValue) instead of target value to prevent flicker
    var shouldAnimate = ((_d = gauge.props.pointer) === null || _d === void 0 ? void 0 : _d.animate) !== false;
    var animateValue = valueLabel.animateValue === true;
    var minValue = gauge.props.minValue;
    // Handle multi-pointer mode
    if (isMultiPointer) {
        if (multiPointerDisplay === 'all') {
            addMultiPointerValueText(gauge, valueLabel, x, y, animateValue && shouldAnimate);
        }
        else {
            // 'primary' - show first pointer value
            var primaryValue = pointers[0].value;
            // Use previous value if animateValue is enabled to prevent flicker
            if (animateValue && shouldAnimate) {
                var prevPointerValue = (_h = (_g = (_f = (_e = gauge.prevProps) === null || _e === void 0 ? void 0 : _e.current) === null || _f === void 0 ? void 0 : _f.pointers) === null || _g === void 0 ? void 0 : _g[0]) === null || _h === void 0 ? void 0 : _h.value;
                primaryValue = prevPointerValue !== null && prevPointerValue !== void 0 ? prevPointerValue : minValue;
            }
            addSingleValueText(gauge, valueLabel, primaryValue, x, y);
        }
        return;
    }
    // Single pointer mode
    var value = gauge.props.value;
    // Use previous value if animateValue is enabled to prevent flicker
    if (animateValue && shouldAnimate) {
        var prevValue = (_k = (_j = gauge.prevProps) === null || _j === void 0 ? void 0 : _j.current) === null || _k === void 0 ? void 0 : _k.value;
        value = prevValue !== null && prevValue !== void 0 ? prevValue : minValue;
    }
    addSingleValueText(gauge, valueLabel, value, x, y);
};
/**
 * Add a single value text to the gauge
 */
var addSingleValueText = function (gauge, valueLabel, value, x, y) {
    var _a, _b;
    var maxDecimalDigits = valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.maxDecimalDigits;
    var floatValue = utils.floatingNumber(value, maxDecimalDigits);
    var arcColor = ((_a = getArcDataByValue(value, gauge)) === null || _a === void 0 ? void 0 : _a.color) || "white";
    // Check if user wants to render custom React content
    if (valueLabel.renderContent) {
        addCustomValueContent(gauge, valueLabel, floatValue, arcColor, x, y);
        return;
    }
    // Standard text rendering
    var text = '';
    if (valueLabel.formatTextValue) {
        text = valueLabel.formatTextValue(floatValue);
    }
    else if (gauge.props.minValue === 0 && gauge.props.maxValue === 100) {
        text = floatValue.toString() + "%";
    }
    else {
        text = floatValue.toString();
    }
    var maxLengthBeforeComputation = 4;
    var textLength = (text === null || text === void 0 ? void 0 : text.length) || 0;
    var fontRatio = textLength > maxLengthBeforeComputation ? maxLengthBeforeComputation / textLength * 1.5 : 1;
    var valueFontSize = (((_b = valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.style) === null || _b === void 0 ? void 0 : _b.fontSize) || '35px');
    var valueTextStyle = __assign({}, (valueLabel.style || {}));
    valueTextStyle.textAnchor = "middle";
    var widthFactor = gauge.props.type == GaugeType.Radial ? 0.003 : 0.003;
    fontRatio = gauge.dimensions.current.width * widthFactor * fontRatio;
    var fontSizeNumber = parseInt(valueFontSize, 10) * fontRatio;
    valueTextStyle.fontSize = fontSizeNumber + "px";
    if (valueLabel.matchColorWithArc)
        valueTextStyle.fill = arcColor;
    addText(text, x, y, gauge, valueTextStyle, CONSTANTS.valueLabelClassname);
};
/**
 * Add multiple pointer values stacked vertically
 */
var addMultiPointerValueText = function (gauge, valueLabel, baseX, baseY, useStartValue) {
    var _a;
    if (useStartValue === void 0) { useStartValue = false; }
    var pointers = gauge.props.pointers;
    if (!pointers || pointers.length === 0)
        return;
    var maxDecimalDigits = valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.maxDecimalDigits;
    var valueFontSize = (((_a = valueLabel === null || valueLabel === void 0 ? void 0 : valueLabel.style) === null || _a === void 0 ? void 0 : _a.fontSize) || '35px');
    var minValue = gauge.props.minValue;
    // Calculate font size based on gauge dimensions
    var widthFactor = gauge.props.type == GaugeType.Radial ? 0.003 : 0.003;
    var fontRatio = gauge.dimensions.current.width * widthFactor;
    var baseFontSizeNumber = parseInt(valueFontSize, 10) * fontRatio;
    // Reduce font size for multiple values to fit better
    var fontSizeReduction = Math.max(0.5, 1 - (pointers.length - 1) * 0.15);
    var fontSize = baseFontSizeNumber * fontSizeReduction;
    var lineHeight = fontSize * 1.3;
    // Center the stack vertically
    var totalHeight = lineHeight * pointers.length;
    var currentY = baseY - totalHeight / 2 + lineHeight / 2;
    pointers.forEach(function (pointer, index) {
        var _a, _b, _c, _d, _e;
        // Use previous value if useStartValue is true (for animateValue feature)
        var displayValue = pointer.value;
        if (useStartValue) {
            var prevPointerValue = (_d = (_c = (_b = (_a = gauge.prevProps) === null || _a === void 0 ? void 0 : _a.current) === null || _b === void 0 ? void 0 : _b.pointers) === null || _c === void 0 ? void 0 : _c[index]) === null || _d === void 0 ? void 0 : _d.value;
            displayValue = prevPointerValue !== null && prevPointerValue !== void 0 ? prevPointerValue : minValue;
        }
        var floatValue = utils.floatingNumber(displayValue, maxDecimalDigits);
        var arcColor = pointer.color || ((_e = getArcDataByValue(pointer.value, gauge)) === null || _e === void 0 ? void 0 : _e.color) || "white";
        var text = '';
        if (valueLabel.formatTextValue) {
            text = valueLabel.formatTextValue(floatValue);
        }
        else if (gauge.props.minValue === 0 && gauge.props.maxValue === 100) {
            text = floatValue.toString() + "%";
        }
        else {
            text = floatValue.toString();
        }
        // Add label if provided
        if (pointer.label) {
            text = "".concat(pointer.label, ": ").concat(text);
        }
        var valueTextStyle = __assign({}, (valueLabel.style || {}));
        valueTextStyle.textAnchor = "middle";
        valueTextStyle.fontSize = fontSize + "px";
        if (valueLabel.matchColorWithArc) {
            valueTextStyle.fill = arcColor;
        }
        addText(text, baseX, currentY, gauge, valueTextStyle, "".concat(CONSTANTS.valueLabelClassname, " multi-value-").concat(index));
        currentY += lineHeight;
    });
};
/**
 * Adds custom React content to the gauge using foreignObject.
 * This allows users to render any React element as the value label.
 *
 * Note: The actual React content is rendered by the main GaugeComponent
 * using the customContent ref. This function just sets up the foreignObject
 * container and stores the render configuration.
 */
var addCustomValueContent = function (gauge, valueLabel, value, arcColor, x, y) {
    var contentWidth = valueLabel.contentWidth || 120;
    var contentHeight = valueLabel.contentHeight || 60;
    // Create foreignObject to embed HTML content in SVG
    var foreignObject = gauge.g.current
        .append("foreignObject")
        .attr("class", CONSTANTS.valueLabelClassname)
        .attr("x", x - contentWidth / 2)
        .attr("y", y - contentHeight / 2)
        .attr("width", contentWidth)
        .attr("height", contentHeight)
        .style("overflow", "visible");
    // Create a container div using standard DOM API for proper React portal compatibility
    // D3's xhtml:div can have namespace issues with React portals
    var foreignObjectNode = foreignObject.node();
    var containerDiv = document.createElement('div');
    containerDiv.className = 'gauge-custom-value-content';
    // Use flex-direction: column to allow block-level elements to stack vertically
    // This ensures h1, hr, p, etc. render correctly with their default block behavior
    containerDiv.style.cssText = 'width: 100%; height: 100%; display: flex; flex-direction: column; align-items: center; justify-content: center; overflow: visible; pointer-events: none;';
    foreignObjectNode.appendChild(containerDiv);
    var container = { node: function () { return containerDiv; } };
    // For simple cases, we can render the content directly
    // The renderContent function returns a React element, but we need to handle it
    // Since we're in D3 context, we'll use a workaround: render to string for static content
    // or store the config for the React component to handle
    // Store the render function and position info on the gauge for React to use
    if (!gauge.customContent) {
        gauge.customContent = { current: {} };
    }
    gauge.customContent.current = {
        containerId: 'gauge-custom-value-content',
        renderContent: valueLabel.renderContent,
        value: value,
        arcColor: arcColor,
    };
    // Get the DOM node and store it for React to render into
    var domNode = container.node();
    if (domNode) {
        registerCustomContentItem(gauge, domNode, valueLabel.renderContent, value, arcColor);
    }
};
var addCustomTickContent = function (gauge, tickValueConfig, value, arcColor, x, y) {
    var contentWidth = tickValueConfig.contentWidth || 60;
    var contentHeight = tickValueConfig.contentHeight || 30;
    var foreignObject = gauge.g.current
        .append("foreignObject")
        .attr("class", CONSTANTS.tickValueClassname)
        .attr("x", x - contentWidth / 2)
        .attr("y", y - contentHeight / 2)
        .attr("width", contentWidth)
        .attr("height", contentHeight)
        .style("overflow", "visible");
    var container = foreignObject
        .append("xhtml:div")
        .attr("class", "gauge-custom-tick-content")
        .style("width", "100%")
        .style("height", "100%")
        .style("display", "flex")
        .style("align-items", "center")
        .style("justify-content", "center")
        .style("overflow", "visible")
        .style("pointer-events", "none");
    var domNode = container.node();
    if (domNode) {
        registerCustomContentItem(gauge, domNode, tickValueConfig.renderContent, value, arcColor);
    }
};
export var clearValueLabel = function (gauge) { return gauge.g.current.selectAll(".".concat(CONSTANTS.valueLabelClassname)).remove(); };
/**
 * Updates the value label text during animation.
 * Called from pointer animation when animateValue is true.
 */
export var updateValueLabelText = function (gauge, currentValue) {
    var _a, _b;
    var labels = gauge.props.labels;
    var valueLabel = labels === null || labels === void 0 ? void 0 : labels.valueLabel;
    if (!valueLabel || valueLabel.hide)
        return;
    var maxDecimalDigits = (_a = valueLabel.maxDecimalDigits) !== null && _a !== void 0 ? _a : 2;
    var floatValue = utils.floatingNumber(currentValue, maxDecimalDigits);
    var text = '';
    if (valueLabel.formatTextValue) {
        text = valueLabel.formatTextValue(floatValue);
    }
    else if (gauge.props.minValue === 0 && gauge.props.maxValue === 100) {
        text = floatValue.toString() + "%";
    }
    else {
        text = floatValue.toString();
    }
    // Update the text element
    var textElement = gauge.g.current.select(".".concat(CONSTANTS.valueLabelClassname, " text"));
    if (!textElement.empty()) {
        textElement.text(text);
        // Update color if matchColorWithArc is enabled
        if (valueLabel.matchColorWithArc) {
            var arcColor = ((_b = getArcDataByValue(currentValue, gauge)) === null || _b === void 0 ? void 0 : _b.color) || "white";
            textElement.style("fill", arcColor);
        }
    }
};
export var clearTicks = function (gauge) {
    var _a;
    // Safety check - g might not be initialized on mobile during deferred render
    if (!((_a = gauge.g.current) === null || _a === void 0 ? void 0 : _a.selectAll))
        return;
    gauge.g.current.selectAll(".".concat(CONSTANTS.tickLineClassname)).remove();
    gauge.g.current.selectAll(".".concat(CONSTANTS.tickValueClassname)).remove();
};
export var calculateAnchorAndAngleByValue = function (value, gauge) {
    var _a;
    var labels = gauge.props.labels;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var valuePercentage = utils.calculatePercentage(minValue, maxValue, value);
    // Use actual angles from gauge dimensions (supports custom angles)
    // D3 angles: 0 = top (12 o'clock), positive = clockwise
    // Convert to degrees for tick line rotation
    var _b = getEffectiveAngles(gauge), d3Start = _b.startAngle, d3End = _b.endAngle;
    var d3Angle = d3Start + valuePercentage * (d3End - d3Start);
    // Convert D3 radians to degrees and subtract 90° to get outward-pointing angle
    // D3 angle 0 (top) -> tick angle -90°/270° (pointing up in screen coords, outward from top)
    // D3 angle π/2 (right) -> tick angle 0° (pointing right, outward)
    // D3 angle -π/2 (left) -> tick angle 180° (pointing left, outward)
    var angle = utils.radToDeg(d3Angle) - 90;
    var isValueLessThanHalf = valuePercentage < 0.5;
    //Values between 40% and 60% are aligned in the middle
    var isValueBetweenTolerance = valuePercentage > CONSTANTS.rangeBetweenCenteredTickValueLabel[0] &&
        valuePercentage < CONSTANTS.rangeBetweenCenteredTickValueLabel[1];
    var tickAnchor = '';
    var isInner = ((_a = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _a === void 0 ? void 0 : _a.type) == "inner";
    if (isValueBetweenTolerance) {
        tickAnchor = "middle";
    }
    else if (isValueLessThanHalf) {
        tickAnchor = isInner ? "start" : "end";
    }
    else {
        tickAnchor = isInner ? "end" : "start";
    }
    // if(valuePercentage > 0.50) angle = angle - 180;
    return { tickAnchor: tickAnchor, angle: angle };
};
var shouldHideTickLine = function (tick, gauge) {
    var _a, _b, _c, _d;
    var labels = gauge.props.labels;
    var defaultHideValue = (_a = defaultTickLabels.defaultTickLineConfig) === null || _a === void 0 ? void 0 : _a.hide;
    var shouldHide = defaultHideValue;
    var defaultHideLineFromUser = (_c = (_b = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _b === void 0 ? void 0 : _b.defaultTickLineConfig) === null || _c === void 0 ? void 0 : _c.hide;
    if (defaultHideLineFromUser != undefined) {
        shouldHide = defaultHideLineFromUser;
    }
    var specificHideValueFromUser = (_d = tick.lineConfig) === null || _d === void 0 ? void 0 : _d.hide;
    if (specificHideValueFromUser != undefined) {
        shouldHide = specificHideValueFromUser;
    }
    return shouldHide;
};
var shouldHideTickValue = function (tick, gauge) {
    var _a, _b, _c, _d;
    var labels = gauge.props.labels;
    var defaultHideValue = (_a = defaultTickLabels.defaultTickValueConfig) === null || _a === void 0 ? void 0 : _a.hide;
    var shouldHide = defaultHideValue;
    var defaultHideValueFromUser = (_c = (_b = labels === null || labels === void 0 ? void 0 : labels.tickLabels) === null || _b === void 0 ? void 0 : _b.defaultTickValueConfig) === null || _c === void 0 ? void 0 : _c.hide;
    if (defaultHideValueFromUser != undefined) {
        shouldHide = defaultHideValueFromUser;
    }
    var specificHideValueFromUser = (_d = tick.valueConfig) === null || _d === void 0 ? void 0 : _d.hide;
    if (specificHideValueFromUser != undefined) {
        shouldHide = specificHideValueFromUser;
    }
    return shouldHide;
};
