// react-gauge-component/dist/lib/GaugeComponent/hooks/pointer.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { easeElastic, easeExpOut } from "d3-ease";
import { interpolateNumber } from "d3-interpolate";
import { drag } from "d3-drag";
import { select } from "d3-selection";
import "d3-transition";
import { PointerType, defaultPointerContext, getPointerLengthByType } from "../types/Pointer";
import { getEffectiveAngles } from "./arc";
import * as utils from "./utils";
import * as arcHooks from "./arc";
import * as labelsHooks from "./labels";
import { GaugeType } from "../types/GaugeComponentProps";
// Global diagnostics tracker (reset per gauge instance in production)
var diagnostics = {
    pointerVanishCount: 0,
    pointerMismatchCount: 0,
    stalePointerCleanups: 0,
    resizeInterruptions: 0,
    animationInterruptions: 0,
    lastVanishTimestamp: null,
    lastMismatchDetails: null,
};
export var getDiagnostics = function () { return (__assign({}, diagnostics)); };
export var resetDiagnostics = function () {
    diagnostics.pointerVanishCount = 0;
    diagnostics.pointerMismatchCount = 0;
    diagnostics.stalePointerCleanups = 0;
    diagnostics.resizeInterruptions = 0;
    diagnostics.animationInterruptions = 0;
    diagnostics.lastVanishTimestamp = null;
    diagnostics.lastMismatchDetails = null;
};
/**
 * Ensure pointer element exists, recreate if missing
 * Tracks vanish events for diagnostics
 */
export var ensurePointerExists = function (gauge) {
    var _a, _b, _c, _d, _e, _f;
    var isMultiPointer = isMultiPointerMode(gauge);
    var shouldHavePointers = hasPointersToRender(gauge);
    // If no pointers should be rendered, always return true (nothing to check)
    if (!shouldHavePointers) {
        return true;
    }
    if (isMultiPointer) {
        var expectedCount = (_b = (_a = gauge.props.pointers) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
        var actualCount = (_d = (_c = gauge.g.current) === null || _c === void 0 ? void 0 : _c.selectAll('.multi-pointer').size()) !== null && _d !== void 0 ? _d : 0;
        if (actualCount < expectedCount) {
            diagnostics.pointerVanishCount++;
            diagnostics.lastVanishTimestamp = Date.now();
            console.warn("[GaugeComponent] Pointer vanished during resize! Expected: ".concat(expectedCount, ", Found: ").concat(actualCount, ". Recreating..."));
            return false;
        }
    }
    else {
        var pointerExists = ((_e = gauge.pointer.current) === null || _e === void 0 ? void 0 : _e.element) &&
            !((_f = gauge.g.current) === null || _f === void 0 ? void 0 : _f.select('.pointer').empty());
        if (!pointerExists) {
            diagnostics.pointerVanishCount++;
            diagnostics.lastVanishTimestamp = Date.now();
            console.warn("[GaugeComponent] Single pointer vanished during resize! Recreating...");
            return false;
        }
    }
    return true;
};
export var drawPointer = function (gauge, resize) {
    var _a, _b, _c, _d, _e, _f, _g;
    if (resize === void 0) { resize = false; }
    // CRITICAL: Check if pointer exists before drawing, recreate if vanished
    if (!ensurePointerExists(gauge)) {
        addPointerElement(gauge);
    }
    var pointer = gauge.props.pointer;
    // Use initialAnimationTriggered flag to handle ResizeObserver firing after prevProps is set
    var isFirstAnimation = !((_a = gauge.initialAnimationTriggered) === null || _a === void 0 ? void 0 : _a.current);
    gauge.pointer.current.context = setupContext(gauge, isFirstAnimation);
    var _h = gauge.pointer.current.context, prevPercent = _h.prevPercent, currentPercent = _h.currentPercent, prevProgress = _h.prevProgress;
    // When resize=true (config change, not value change), draw directly at currentPercent
    var useCurrentPercent = resize && !isFirstAnimation;
    // Initialize pointer for all types except Grafana (which uses arc fill by default)
    var isGrafana = gauge.props.type == GaugeType.Grafana;
    var userExplicitlyConfiguredPointer = ((_b = gauge.originalProps) === null || _b === void 0 ? void 0 : _b.pointer) !== undefined;
    var showPointerForGrafana = isGrafana && userExplicitlyConfiguredPointer && !pointer.hide;
    // Only init pointer on first animation OR on resize when animation is enabled
    var shouldInitPointer = isFirstAnimation || (resize && pointer.animate !== false);
    if (shouldInitPointer && (!isGrafana || showPointerForGrafana)) {
        initPointer(gauge, useCurrentPercent);
    }
    // Skip animation if user is currently dragging the pointer
    var isDragging = ((_c = gauge.isDragging) === null || _c === void 0 ? void 0 : _c.current) === true;
    var shouldAnimate = (!resize || isFirstAnimation) && pointer.animate && !isDragging;
    if (shouldAnimate) {
        // Mark that initial animation has been triggered to prevent ResizeObserver from restarting
        if (gauge.initialAnimationTriggered) {
            gauge.initialAnimationTriggered.current = true;
        }
        // Mark animation as in progress
        if (gauge.animationInProgress) {
            gauge.animationInProgress.current = true;
        }
        // For Grafana type without pointer, animate the doughnut (arc fill animation)
        // For other types or Grafana with pointer, animate the pointer element
        var animationTarget = (isGrafana && !showPointerForGrafana)
            ? gauge.doughnut.current
            : (_d = gauge.pointer.current) === null || _d === void 0 ? void 0 : _d.element;
        // Safety check - target might be null during rapid config changes
        if (!animationTarget) {
            if (gauge.animationInProgress) {
                gauge.animationInProgress.current = false;
            }
            return;
        }
        // CRITICAL: Interrupt any existing animation before starting new one
        animationTarget.interrupt();
        // Use the last rendered progress as starting point for smooth mid-animation transitions
        // prevProgress is updated on every animation frame, so it reflects current visual position
        // Use prevProgress if it exists (even if 0), otherwise fall back to prevPercent
        var lastRenderedProgress = (_f = (_e = gauge.pointer.current) === null || _e === void 0 ? void 0 : _e.context) === null || _f === void 0 ? void 0 : _f.prevProgress;
        var hasLastRenderedProgress = lastRenderedProgress !== undefined && !isFirstAnimation;
        var animationStartPercent_1 = hasLastRenderedProgress ? lastRenderedProgress : prevPercent;
        // FPS limiting - calculate minimum time between frames
        var maxFps = (_g = pointer.maxFps) !== null && _g !== void 0 ? _g : 60;
        var minFrameTime_1 = maxFps > 0 ? 1000 / maxFps : 0;
        var lastFrameTime_1 = 0;
        // For first animation: use configured delay and duration for nice intro
        // For subsequent animations: no delay, use configured duration (user controls responsiveness)
        var effectiveDelay = isFirstAnimation ? pointer.animationDelay : 0;
        var effectiveDuration = pointer.animationDuration;
        animationTarget
            .transition()
            .delay(effectiveDelay)
            .ease(pointer.elastic ? easeElastic : easeExpOut)
            .duration(effectiveDuration)
            .tween("progress", function () {
            var currentInterpolatedPercent = interpolateNumber(animationStartPercent_1, currentPercent);
            return function (percentOfPercent) {
                var _a, _b;
                // FPS limiting - skip frame if not enough time has passed
                var now = performance.now();
                if (minFrameTime_1 > 0 && (now - lastFrameTime_1) < minFrameTime_1) {
                    // Skip this frame, but still update on final frame
                    if (percentOfPercent < 0.99)
                        return;
                }
                lastFrameTime_1 = now;
                var progress = currentInterpolatedPercent(percentOfPercent);
                if (isProgressValid(progress, prevProgress, gauge)) {
                    // Always update Grafana arc fill
                    if (isGrafana) {
                        arcHooks.updateGrafanaArc(gauge, progress);
                    }
                    // Update pointer if not Grafana, or if Grafana with pointer shown
                    if (!isGrafana || showPointerForGrafana) {
                        updatePointer(progress, gauge);
                    }
                    // Update value label in real-time if animateValue is enabled
                    if ((_b = (_a = gauge.props.labels) === null || _a === void 0 ? void 0 : _a.valueLabel) === null || _b === void 0 ? void 0 : _b.animateValue) {
                        var minValue = gauge.props.minValue;
                        var maxValue = gauge.props.maxValue;
                        var currentValue = minValue + progress * (maxValue - minValue);
                        labelsHooks.updateValueLabelText(gauge, currentValue);
                    }
                }
                gauge.pointer.current.context.prevProgress = progress;
            };
        })
            .on("end", function () {
            var _a;
            // Mark animation as complete
            if (gauge.animationInProgress) {
                gauge.animationInProgress.current = false;
            }
            // Apply pending config changes that were queued during animation
            if ((_a = gauge.pendingConfigChange) === null || _a === void 0 ? void 0 : _a.current) {
                gauge.pendingConfigChange.current = false;
                // Use requestAnimationFrame to avoid blocking and ensure clean state
                requestAnimationFrame(function () {
                    // Import dynamically to avoid circular dependency
                    var chartHooks = require('./chart');
                    chartHooks.initChart(gauge, false);
                });
            }
        });
    }
    else {
        // Mark initial animation as triggered even when animation is disabled
        // so subsequent updates know it's not the first render
        if (isFirstAnimation && gauge.initialAnimationTriggered) {
            gauge.initialAnimationTriggered.current = true;
        }
        // For Grafana, always update the arc fill
        if (isGrafana) {
            arcHooks.updateGrafanaArc(gauge, currentPercent);
        }
        // Update pointer if not Grafana, or if Grafana with pointer shown
        if (!isGrafana || showPointerForGrafana) {
            updatePointer(currentPercent, gauge);
        }
    }
};
export var setupContext = function (gauge, isFirstAnimation) {
    var _a, _b, _c, _d;
    if (isFirstAnimation === void 0) { isFirstAnimation = false; }
    var value = gauge.props.value;
    var pointer = gauge.props.pointer;
    var pointerLength = pointer.length;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var pointerPath = gauge.pointer.current.context.pointerPath;
    var pointerRadius = getPointerRadius(gauge);
    var length = pointerLength;
    var typesWithPath = [PointerType.Needle, PointerType.Arrow];
    // CRITICAL: Preserve the last rendered progress from previous context
    // This is updated every animation frame and represents actual visual position
    // Without this, animation always starts from 0 instead of current position
    var existingPrevProgress = (_b = (_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.context) === null || _b === void 0 ? void 0 : _b.prevProgress;
    // Handle prevValue properly - use nullish coalescing to allow 0 values
    // CRITICAL: On first animation, always use minValue so pointer animates from start
    // This fixes the issue where prevProps.current.value is already set to current value
    // before the ResizeObserver fires, causing no animation on first load
    var prevValue = isFirstAnimation ? minValue : ((_d = (_c = gauge.prevProps) === null || _c === void 0 ? void 0 : _c.current.value) !== null && _d !== void 0 ? _d : minValue);
    var pointerContext = {
        centerPoint: [0, -pointerRadius / 2],
        pointerRadius: getPointerRadius(gauge),
        pathLength: gauge.dimensions.current.outerRadius * length,
        currentPercent: utils.calculatePercentage(minValue, maxValue, value),
        prevPercent: utils.calculatePercentage(minValue, maxValue, prevValue),
        // Preserve existing prevProgress if available, otherwise use 0 for first animation
        prevProgress: isFirstAnimation ? 0 : (existingPrevProgress !== null && existingPrevProgress !== void 0 ? existingPrevProgress : 0),
        pathStr: "",
        shouldDrawPath: typesWithPath.includes(pointer.type),
        prevColor: ""
    };
    return pointerContext;
};
var initPointer = function (gauge, useCurrentPercent) {
    var _a, _b;
    if (useCurrentPercent === void 0) { useCurrentPercent = false; }
    // Safety check - element might be null during rapid config changes
    if (!((_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.element) || !((_b = gauge.pointer.current) === null || _b === void 0 ? void 0 : _b.context))
        return;
    // CRITICAL: Clear existing pointer children before creating new ones
    // This ensures Arrow/Blob/Needle are properly recreated when props change
    gauge.pointer.current.element.selectAll("*").remove();
    gauge.pointer.current.path = null;
    var value = gauge.props.value;
    var pointer = gauge.props.pointer;
    var _c = gauge.pointer.current.context, shouldDrawPath = _c.shouldDrawPath, centerPoint = _c.centerPoint, pointerRadius = _c.pointerRadius, pathStr = _c.pathStr, currentPercent = _c.currentPercent, prevPercent = _c.prevPercent, pathLength = _c.pathLength;
    // Get the initial color based on current value - this makes pointer color match arc by default
    var initialColor = pointer.color || arcHooks.getColorByPercentage(currentPercent, gauge);
    // Use currentPercent when reinitializing due to config changes (not value changes)
    // This prevents the pointer from jumping from old value to new value
    // Note: Use nullish coalescing (??) instead of || to handle prevPercent=0 correctly
    var startPercent = useCurrentPercent ? currentPercent : (prevPercent !== null && prevPercent !== void 0 ? prevPercent : 0);
    if (shouldDrawPath) {
        gauge.pointer.current.context.pathStr = calculatePointerPath(gauge, startPercent);
        var pathElement = gauge.pointer.current.element.append("path")
            .attr("d", gauge.pointer.current.context.pathStr)
            .attr("fill", initialColor);
        // Add stroke/border if configured
        var strokeWidth = pointer.strokeWidth || 0;
        if (strokeWidth > 0) {
            var strokeColor = pointer.strokeColor || 'rgba(255, 255, 255, 0.8)';
            pathElement
                .attr("stroke", strokeColor)
                .attr("stroke-width", strokeWidth)
                .attr("stroke-linejoin", "round");
        }
        gauge.pointer.current.path = pathElement;
        // Add grab handle at pointer tip if onValueChange is provided and not hidden
        // Note: The handle will be raised to top after all elements are rendered in drawPointer
        if (gauge.props.onValueChange && !pointer.hideGrabHandle) {
            var tipPosition = calculatePointerTipPosition(gauge, startPercent);
            var handleRadius = Math.max(6, pointerRadius * 0.8);
            // Append to the main g element (not pointer element) so it renders on top of everything
            gauge.g.current
                .append("circle")
                .attr("class", "pointer-grab-handle")
                .attr("cx", tipPosition.x)
                .attr("cy", tipPosition.y)
                .attr("r", handleRadius)
                .attr("fill", "rgba(255, 255, 255, 0.3)")
                .attr("stroke", "#fff")
                .attr("stroke-width", 2)
                .style("cursor", "grab");
        }
    }
    //Add a circle at the center (base of pointer)
    if (pointer.type == PointerType.Needle) {
        var needleBaseCircle = gauge.pointer.current.element
            .append("circle")
            .attr("cx", 0)
            .attr("cy", 0)
            .attr("r", pointerRadius)
            .attr("fill", pointer.baseColor || initialColor);
        // Add stroke to needle base circle if configured
        var strokeWidth = pointer.strokeWidth || 0;
        if (strokeWidth > 0) {
            var strokeColor = pointer.strokeColor || 'rgba(255, 255, 255, 0.8)';
            needleBaseCircle
                .attr("stroke", strokeColor)
                .attr("stroke-width", strokeWidth);
        }
    }
    else if (pointer.type == PointerType.Blob) {
        // For blob, stroke color matches arc color by default, or use custom strokeColor
        var arcColor = arcHooks.getColorByPercentage(currentPercent, gauge);
        var strokeColor = pointer.strokeColor || arcColor;
        var strokeWidth = pointer.strokeWidth !== undefined ? pointer.strokeWidth : 8;
        // Blob circle centered at (0,0) so translation places it exactly on target
        gauge.pointer.current.element
            .append("circle")
            .attr("cx", 0)
            .attr("cy", 0)
            .attr("r", pointerRadius)
            .attr("fill", pointer.baseColor)
            .attr("stroke", strokeColor)
            .attr("stroke-width", strokeWidth * pointerRadius / 10);
        gauge.pointer.current.context.prevColor = arcColor;
        // For blob, the blob itself is the grab handle - just add cursor style
        if (gauge.props.onValueChange) {
            gauge.pointer.current.element.select("circle").style("cursor", "grab");
        }
    }
    //Translate the pointer starting point of the arc
    setPointerPosition(pointerRadius, startPercent, gauge);
};
var updatePointer = function (percentage, gauge) {
    var _a, _b, _c;
    // Safety check - element might be null during rapid config changes
    if (!((_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.element) || !((_b = gauge.pointer.current) === null || _b === void 0 ? void 0 : _b.context))
        return;
    var pointer = gauge.props.pointer;
    var _d = gauge.pointer.current.context, pointerRadius = _d.pointerRadius, shouldDrawPath = _d.shouldDrawPath, prevColor = _d.prevColor;
    setPointerPosition(pointerRadius, percentage, gauge);
    if (shouldDrawPath && gauge.pointer.current.path) {
        gauge.pointer.current.path.attr("d", calculatePointerPath(gauge, percentage));
        // Update grab handle position if it exists (grab handle is on g.current, not pointer element)
        var grabHandle = (_c = gauge.g.current) === null || _c === void 0 ? void 0 : _c.select(".pointer-grab-handle");
        if (grabHandle && !grabHandle.empty()) {
            var tipPosition = calculatePointerTipPosition(gauge, percentage);
            grabHandle.attr("cx", tipPosition.x).attr("cy", tipPosition.y);
        }
    }
    // Update arrow color - use fixed color if set, otherwise match arc color
    if (pointer.type == PointerType.Arrow && gauge.pointer.current.path) {
        var currentColor = pointer.color || arcHooks.getColorByPercentage(percentage, gauge);
        gauge.pointer.current.path.attr("fill", currentColor);
    }
    // Update needle color as well
    if (pointer.type == PointerType.Needle && shouldDrawPath && gauge.pointer.current.path) {
        var currentColor = pointer.color || arcHooks.getColorByPercentage(percentage, gauge);
        gauge.pointer.current.path.attr("fill", currentColor);
    }
    if (pointer.type == PointerType.Blob && gauge.pointer.current.element) {
        // Use getColorByPercentage which handles both gradient and non-gradient modes
        var arcColor = arcHooks.getColorByPercentage(percentage, gauge);
        // Use custom strokeColor if provided, otherwise match arc color
        var currentColor = pointer.strokeColor || arcColor;
        var shouldChangeColor = arcColor != prevColor;
        if (shouldChangeColor && !pointer.strokeColor) {
            gauge.pointer.current.element.select("circle").attr("stroke", currentColor);
        }
        var strokeWidth = pointer.strokeWidth !== undefined ? pointer.strokeWidth : 8;
        gauge.pointer.current.element.select("circle").attr("stroke-width", strokeWidth * pointerRadius / 10);
        gauge.pointer.current.context.prevColor = arcColor;
    }
};
var setPointerPosition = function (pointerRadius, progress, gauge) {
    var _a;
    var pointer = gauge.props.pointer;
    var pointerType = pointer.type;
    var dimensions = gauge.dimensions;
    var value = utils.getCurrentGaugeValueByPercentage(progress, gauge);
    var pointers = (_a = {},
        _a[PointerType.Needle] = function () {
            // Set needle position to center (origin, since g is already centered)
            translatePointer(0, 0, gauge);
        },
        _a[PointerType.Arrow] = function () {
            var _a;
            // Position arrow based on arrowOffset (0 = center, 1 = outer edge of arc)
            var arrowOffset = (_a = pointer.arrowOffset) !== null && _a !== void 0 ? _a : 0.72;
            var innerR = gauge.dimensions.current.innerRadius;
            var outerR = gauge.dimensions.current.outerRadius;
            // Arrow offset is relative to inner radius toward outer
            var targetRadius = innerR * arrowOffset;
            // Use the SAME angles as the arc drawing
            var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
            var angle = startAngle + progress * (endAngle - startAngle);
            // Convert d3 angle to x,y coordinates
            var x = targetRadius * Math.sin(angle);
            var y = -targetRadius * Math.cos(angle);
            translatePointer(x, y, gauge);
        },
        _a[PointerType.Blob] = function () {
            var _a;
            // Position blob based on blobOffset (0 = inner edge, 0.5 = center, 1 = outer edge)
            var blobOffset = (_a = pointer.blobOffset) !== null && _a !== void 0 ? _a : 0.5;
            var innerR = gauge.dimensions.current.innerRadius;
            var outerR = gauge.dimensions.current.outerRadius;
            // Interpolate between inner and outer radius based on offset
            var targetRadius = innerR + (outerR - innerR) * blobOffset;
            // Use the SAME angles as the arc drawing
            // These are d3 angles: 0 at top, positive clockwise
            var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
            var angle = startAngle + progress * (endAngle - startAngle);
            // Convert d3 angle to x,y coordinates
            // In d3 convention: 0 is at top, angles go clockwise
            // x = radius * sin(angle), y = -radius * cos(angle)
            var x = targetRadius * Math.sin(angle);
            var y = -targetRadius * Math.cos(angle);
            translatePointer(x, y, gauge);
        },
        _a);
    return pointers[pointerType]();
};
/**
 * Validates if progress update should trigger a DOM update.
 * Uses animationThreshold from pointer props to control update frequency.
 */
var isProgressValid = function (currentPercent, prevPercent, gauge) {
    var _a;
    var pointer = gauge.props.pointer;
    // Use configurable threshold (default 0.001 for backward compatibility)
    var threshold = (_a = pointer.animationThreshold) !== null && _a !== void 0 ? _a : 0.001;
    //Avoid unnecessary re-rendering (when progress is too small) but allow the pointer to reach the final value
    var overFlow = currentPercent > 1 || currentPercent < 0;
    var tooSmallValue = Math.abs(currentPercent - prevPercent) < threshold;
    var sameValueAsBefore = currentPercent == prevPercent;
    return !tooSmallValue && !sameValueAsBefore && !overFlow;
};
var calculatePointerPath = function (gauge, percent) {
    var _a = gauge.pointer.current.context, pointerRadius = _a.pointerRadius, pathLength = _a.pathLength;
    // Use actual angles from gauge dimensions (supports custom angles)
    // D3 angle convention: 0 = top (12 o'clock), positive = clockwise
    var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
    var d3Angle = startAngle + percent * (endAngle - startAngle);
    // Calculate needle tip position using D3 angle convention (same as arc)
    // x = radius * sin(angle), y = -radius * cos(angle)
    var tipX = pathLength * Math.sin(d3Angle);
    var tipY = -pathLength * Math.cos(d3Angle);
    // Calculate base points perpendicular to the needle direction
    // The base is at the center (0, 0), with points spread perpendicular to the needle
    var perpAngle = d3Angle + Math.PI / 2; // perpendicular angle
    var baseOffset = pointerRadius;
    var leftX = baseOffset * Math.sin(perpAngle);
    var leftY = -baseOffset * Math.cos(perpAngle);
    var rightX = -baseOffset * Math.sin(perpAngle);
    var rightY = baseOffset * Math.cos(perpAngle);
    var pathStr = "M ".concat(leftX, " ").concat(leftY, " L ").concat(tipX, " ").concat(tipY, " L ").concat(rightX, " ").concat(rightY);
    return pathStr;
};
/**
 * Calculate the position of the pointer tip for the grab handle
 */
var calculatePointerTipPosition = function (gauge, percent) {
    var _a;
    var pointer = gauge.props.pointer;
    var pointerType = pointer.type;
    var pathLength = gauge.pointer.current.context.pathLength;
    var innerR = gauge.dimensions.current.innerRadius;
    // Use the SAME angles as the arc drawing (D3 convention)
    var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
    var d3Angle = startAngle + percent * (endAngle - startAngle);
    // For Arrow type, calculate position based on arrow offset + local tip
    if (pointerType === PointerType.Arrow) {
        var arrowOffset = (_a = pointer.arrowOffset) !== null && _a !== void 0 ? _a : 0.72;
        var targetRadius = innerR * arrowOffset;
        // Arrow translation position
        var transX = targetRadius * Math.sin(d3Angle);
        var transY = -targetRadius * Math.cos(d3Angle);
        // Local tip position (arrow path tip)
        var localTipX = pathLength * Math.sin(d3Angle);
        var localTipY = -pathLength * Math.cos(d3Angle);
        return {
            x: transX + localTipX,
            y: transY + localTipY,
        };
    }
    // For Needle type, tip is at pathLength from center along the angle
    return {
        x: pathLength * Math.sin(d3Angle),
        y: -pathLength * Math.cos(d3Angle),
    };
};
var getPointerRadius = function (gauge) {
    var pointer = gauge.props.pointer;
    var pointerWidth = pointer.width;
    return pointerWidth * (gauge.dimensions.current.width / 500);
};
export var translatePointer = function (x, y, gauge) {
    var _a;
    if (!((_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.element))
        return;
    gauge.pointer.current.element.attr("transform", "translate(" + x + ", " + y + ")");
};
export var addPointerElement = function (gauge) { return gauge.pointer.current.element = gauge.g.current.append("g").attr("class", "pointer"); };
export var clearPointerElement = function (gauge) {
    var _a, _b;
    // Safety check - element might not exist in multi-pointer mode
    if ((_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.element) {
        gauge.pointer.current.element.selectAll("*").remove();
    }
    // Also remove grab handle which is on g.current
    (_b = gauge.g.current) === null || _b === void 0 ? void 0 : _b.select(".pointer-grab-handle").remove();
};
/**
 * Calculate value from mouse/touch position on the gauge arc
 * @param gauge The gauge instance
 * @param clientX Mouse/touch X coordinate
 * @param clientY Mouse/touch Y coordinate
 * @returns The calculated value based on position
 */
export var getValueFromPosition = function (gauge, clientX, clientY) {
    var svgElement = gauge.svg.current.node();
    if (!svgElement)
        return gauge.props.value;
    var rect = svgElement.getBoundingClientRect();
    var viewBoxAttr = gauge.svg.current.attr("viewBox");
    if (!viewBoxAttr)
        return gauge.props.value;
    var viewBox = viewBoxAttr.split(" ").map(Number);
    // viewBox = [x, y, width, height]
    // Convert client coordinates to SVG viewBox coordinates
    var scaleX = viewBox[2] / rect.width;
    var scaleY = viewBox[3] / rect.height;
    var svgX = (clientX - rect.left) * scaleX + viewBox[0];
    var svgY = (clientY - rect.top) * scaleY + viewBox[1];
    // Get gauge center from the current layout
    // The g element is translated to gaugeCenter, so that's our pivot point
    var layout = gauge.prevGSize.current;
    if (!(layout === null || layout === void 0 ? void 0 : layout.gaugeCenter))
        return gauge.props.value;
    var centerX = layout.gaugeCenter.x;
    var centerY = layout.gaugeCenter.y;
    // Calculate vector from center to mouse position
    var dx = svgX - centerX;
    var dy = svgY - centerY;
    // D3 arc uses angles where:
    //   0 = TOP (12 o'clock), positive angles go CLOCKWISE
    //   -PI/2 = LEFT (9 o'clock)
    //   PI/2 = RIGHT (3 o'clock)
    //   PI or -PI = BOTTOM (6 o'clock)
    //
    // To convert from mouse (dx, dy) to D3 arc angle:
    // atan2(dx, -dy) gives angle from top, clockwise positive
    var angle = Math.atan2(dx, -dy);
    // Get the actual gauge angles
    var _a = getEffectiveAngles(gauge), startAngle = _a.startAngle, endAngle = _a.endAngle;
    // Calculate the dead zone (where there's no arc)
    // The arc goes from startAngle to endAngle
    // The dead zone is from endAngle to startAngle (going the long way around)
    var arcSpan = endAngle - startAngle;
    var deadZoneSpan = 2 * Math.PI - arcSpan;
    var deadZoneCenter = endAngle + deadZoneSpan / 2;
    // Normalize angle to be relative to deadZoneCenter for easier comparison
    var relativeAngle = angle - deadZoneCenter;
    // Normalize to -PI to PI range
    while (relativeAngle > Math.PI)
        relativeAngle -= 2 * Math.PI;
    while (relativeAngle < -Math.PI)
        relativeAngle += 2 * Math.PI;
    // If in dead zone (within deadZoneSpan/2 of deadZoneCenter), snap to nearest edge
    if (Math.abs(relativeAngle) < deadZoneSpan / 2) {
        // Snap to nearest valid edge
        angle = relativeAngle < 0 ? endAngle : startAngle;
    }
    // Clamp angle to valid range
    angle = Math.max(startAngle, Math.min(endAngle, angle));
    // Calculate percentage (0 to 1)
    var percentage = (angle - startAngle) / (endAngle - startAngle);
    // Convert percentage to value
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var value = minValue + percentage * (maxValue - minValue);
    // Clamp value to min/max
    return Math.max(minValue, Math.min(maxValue, value));
};
/**
 * Set up drag behavior for the pointer element
 * This allows users to grab and drag the pointer to set values
 */
export var setupPointerDrag = function (gauge) {
    var onValueChange = gauge.props.onValueChange;
    if (!onValueChange)
        return; // Only enable drag if callback is provided
    var pointerElement = gauge.pointer.current.element;
    if (!pointerElement)
        return;
    // Also make the arc draggable for easier interaction
    var arcElement = gauge.doughnut.current;
    // Track if we've actually moved during drag (to distinguish from click)
    var hasMoved = false;
    var startX = 0;
    var startY = 0;
    var moveThreshold = 3; // Minimum pixels to move before considering it a drag
    var handleDragStart = function (event) {
        // Stop any ongoing animations
        pointerElement.interrupt();
        if (arcElement)
            arcElement.interrupt();
        // Mark as dragging to skip animation during drag
        if (gauge.isDragging) {
            gauge.isDragging.current = true;
        }
        // Record start position
        hasMoved = false;
        startX = event.sourceEvent.clientX;
        startY = event.sourceEvent.clientY;
    };
    var handleDrag = function (event) {
        var _a;
        var clientX = event.sourceEvent.clientX;
        var clientY = event.sourceEvent.clientY;
        // Check if we've moved enough to consider it a drag
        var dx = Math.abs(clientX - startX);
        var dy = Math.abs(clientY - startY);
        if (dx > moveThreshold || dy > moveThreshold) {
            hasMoved = true;
        }
        // Only update value if we've actually moved
        if (hasMoved) {
            var value = getValueFromPosition(gauge, clientX, clientY);
            // Update pointer position directly during drag (no animation)
            var minValue = gauge.props.minValue;
            var maxValue = gauge.props.maxValue;
            var percent = (value - minValue) / (maxValue - minValue);
            updatePointer(percent, gauge);
            // Update Grafana arc if applicable
            if (gauge.props.type === GaugeType.Grafana) {
                arcHooks.updateGrafanaArc(gauge, percent);
            }
            // Update value label
            labelsHooks.updateValueLabelText(gauge, value);
            // Track the current progress for smooth transition after drag
            if ((_a = gauge.pointer.current) === null || _a === void 0 ? void 0 : _a.context) {
                gauge.pointer.current.context.prevProgress = percent;
            }
            // Notify parent of value change
            onValueChange(value);
        }
    };
    var handleDragEnd = function () {
        // Mark drag as complete
        if (gauge.isDragging) {
            gauge.isDragging.current = false;
        }
    };
    var dragBehavior = drag()
        .on("start", handleDragStart)
        .on("drag", handleDrag)
        .on("end", handleDragEnd);
    // Apply drag to pointer
    pointerElement.call(dragBehavior);
    pointerElement.style("cursor", "grab");
    // Apply drag to grab handle (which is on g.current)
    var grabHandle = gauge.g.current.select(".pointer-grab-handle");
    if (!grabHandle.empty()) {
        grabHandle.call(dragBehavior);
    }
    // Also apply drag to the arc for easier interaction
    if (arcElement) {
        arcElement.call(dragBehavior);
        arcElement.style("cursor", "pointer");
    }
};
/**
 * Set up click-to-set behavior on the arc
 * Allows users to click anywhere on the arc to set the value
 */
export var setupArcClick = function (gauge) {
    // Note: Click behavior is now handled separately from drag
    // The arc click will set value on single click (no drag)
    var onValueChange = gauge.props.onValueChange;
    if (!onValueChange)
        return;
    var arcElement = gauge.doughnut.current;
    if (!arcElement)
        return;
    // Use mouseup instead of click to avoid conflict with drag
    // Only fire if it was a quick click (not a drag)
    var mouseDownTime = 0;
    var clickThreshold = 200; // ms - if mouseup happens within this time, treat as click
    arcElement.on("mousedown.click", function () {
        mouseDownTime = Date.now();
    });
    arcElement.on("mouseup.click", function (event) {
        var elapsed = Date.now() - mouseDownTime;
        if (elapsed < clickThreshold) {
            var value = getValueFromPosition(gauge, event.clientX, event.clientY);
            onValueChange(value);
        }
    });
};
// ============================================================================
// MULTI-POINTER SUPPORT
// ============================================================================
/**
 * Check if gauge is in multi-pointer mode
 * Returns true only if pointers array exists AND has at least one pointer
 */
export var isMultiPointerMode = function (gauge) {
    return Array.isArray(gauge.props.pointers) && gauge.props.pointers.length > 0;
};
/**
 * Check if gauge has any pointers to draw
 */
export var hasPointersToRender = function (gauge) {
    var _a;
    if (Array.isArray(gauge.props.pointers)) {
        return gauge.props.pointers.length > 0;
    }
    return !((_a = gauge.props.pointer) === null || _a === void 0 ? void 0 : _a.hide);
};
/**
 * Validate that rendered pointer count matches expected count from props
 * Logs warning and corrects if mismatch detected
 */
export var validatePointerCount = function (gauge) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    var expectedCount = (_b = (_a = gauge.props.pointers) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : (((_c = gauge.props.pointer) === null || _c === void 0 ? void 0 : _c.hide) ? 0 : 1);
    var renderedMultiPointers = (_e = (_d = gauge.g.current) === null || _d === void 0 ? void 0 : _d.selectAll('.multi-pointer').size()) !== null && _e !== void 0 ? _e : 0;
    var renderedSinglePointer = (_g = (_f = gauge.g.current) === null || _f === void 0 ? void 0 : _f.selectAll('.pointer').size()) !== null && _g !== void 0 ? _g : 0;
    var actualCount = ((_h = gauge.props.pointers) === null || _h === void 0 ? void 0 : _h.length) ? renderedMultiPointers : renderedSinglePointer;
    if (actualCount !== expectedCount) {
        console.warn("[GaugeComponent] Pointer count mismatch detected! Expected: ".concat(expectedCount, ", Rendered: ").concat(actualCount, ". Auto-correcting..."));
        return { valid: false, expected: expectedCount, actual: actualCount };
    }
    return { valid: true, expected: expectedCount, actual: actualCount };
};
/**
 * Draw all pointers in multi-pointer mode
 */
export var drawMultiPointers = function (gauge, resize) {
    var _a, _b, _c, _d, _e, _f;
    if (resize === void 0) { resize = false; }
    var pointers = gauge.props.pointers;
    if (!pointers || pointers.length === 0)
        return;
    var isGrafana = gauge.props.type === GaugeType.Grafana;
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    // Initialize multiPointers array if needed
    if (!((_a = gauge.multiPointers) === null || _a === void 0 ? void 0 : _a.current)) {
        gauge.multiPointers.current = [];
    }
    if (!((_b = gauge.multiPointerAnimationTriggered) === null || _b === void 0 ? void 0 : _b.current)) {
        gauge.multiPointerAnimationTriggered.current = [];
    }
    // VALIDATION: Check for stale pointers and clean up
    var currentRenderedCount = (_d = (_c = gauge.g.current) === null || _c === void 0 ? void 0 : _c.selectAll('.multi-pointer').size()) !== null && _d !== void 0 ? _d : 0;
    if (currentRenderedCount > pointers.length) {
        console.warn("[GaugeComponent] Stale multi-pointers detected! Rendered: ".concat(currentRenderedCount, ", Expected: ").concat(pointers.length, ". Cleaning up..."));
        // Remove excess pointer elements
        (_e = gauge.g.current) === null || _e === void 0 ? void 0 : _e.selectAll('.multi-pointer').each(function (d, i) {
            if (i >= pointers.length) {
                select(this).remove();
            }
        });
        // Trim the multiPointers array
        if ((_f = gauge.multiPointers) === null || _f === void 0 ? void 0 : _f.current) {
            gauge.multiPointers.current = gauge.multiPointers.current.slice(0, pointers.length);
        }
    }
    // Ensure we have the right number of animation triggered flags
    while (gauge.multiPointerAnimationTriggered.current.length < pointers.length) {
        gauge.multiPointerAnimationTriggered.current.push(false);
    }
    // Trim excess flags
    if (gauge.multiPointerAnimationTriggered.current.length > pointers.length) {
        gauge.multiPointerAnimationTriggered.current = gauge.multiPointerAnimationTriggered.current.slice(0, pointers.length);
    }
    // Draw each pointer
    pointers.forEach(function (pointerConfig, index) {
        drawSingleMultiPointer(gauge, pointerConfig, index, resize, minValue, maxValue, isGrafana);
    });
    // For Grafana, use primary (first) pointer value for arc fill
    if (isGrafana && pointers.length > 0) {
        var primaryPercent = utils.calculatePercentage(minValue, maxValue, pointers[0].value);
        arcHooks.updateGrafanaArc(gauge, primaryPercent);
    }
};
/**
 * Draw a single pointer in multi-pointer mode
 */
var drawSingleMultiPointer = function (gauge, pointerConfig, index, resize, minValue, maxValue, isGrafana) {
    var _a, _b, _c, _d, _e, _f, _g;
    var defaultPointer = require("../types/Pointer").defaultPointer;
    // Merge with defaults
    var pointer = __assign(__assign({}, defaultPointer), pointerConfig);
    if (pointer.hide)
        return;
    var currentPercent = utils.calculatePercentage(minValue, maxValue, pointer.value);
    var isFirstAnimation = !((_a = gauge.multiPointerAnimationTriggered) === null || _a === void 0 ? void 0 : _a.current[index]);
    // Get or create pointer ref
    var pointerRef = gauge.multiPointers.current[index];
    if (!pointerRef) {
        pointerRef = {
            element: null,
            path: null,
            context: __assign({}, defaultPointerContext),
            index: index,
            animationInProgress: false
        };
        gauge.multiPointers.current[index] = pointerRef;
    }
    // Check if element actually exists in DOM (might have been orphaned)
    var elementExistsInDOM = pointerRef.element && !((_b = gauge.g.current) === null || _b === void 0 ? void 0 : _b.select(".multi-pointer-".concat(index)).empty());
    // Setup context for this pointer
    var pointerRadius = getPointerRadiusForConfig(gauge, pointer);
    // Use type-specific default length if user didn't explicitly provide one,
    // otherwise respect user's explicit length value
    var length = pointerConfig.length !== undefined
        ? pointerConfig.length
        : getPointerLengthByType(pointer.type);
    var typesWithPath = [PointerType.Needle, PointerType.Arrow];
    // Get previous value from prevProps if available
    var prevPointers = (_d = (_c = gauge.prevProps) === null || _c === void 0 ? void 0 : _c.current) === null || _d === void 0 ? void 0 : _d.pointers;
    var prevValue = (_f = (_e = prevPointers === null || prevPointers === void 0 ? void 0 : prevPointers[index]) === null || _e === void 0 ? void 0 : _e.value) !== null && _f !== void 0 ? _f : minValue;
    var prevPercent = utils.calculatePercentage(minValue, maxValue, prevValue);
    pointerRef.context = {
        centerPoint: [0, -pointerRadius / 2],
        pointerRadius: pointerRadius,
        pathLength: gauge.dimensions.current.outerRadius * length,
        currentPercent: currentPercent,
        prevPercent: prevPercent,
        prevProgress: 0,
        pathStr: "",
        shouldDrawPath: typesWithPath.includes(pointer.type),
        prevColor: ((_g = pointerRef.context) === null || _g === void 0 ? void 0 : _g.prevColor) || ""
    };
    var useCurrentPercent = resize && !isFirstAnimation;
    // Only init pointer if it's first animation OR resize with animation enabled
    // BUT skip if element already exists in DOM (prevents duplicates on value-only changes)
    var shouldInitPointer = (isFirstAnimation || (resize && pointer.animate !== false)) && !elementExistsInDOM;
    // Create or update pointer element
    if (shouldInitPointer) {
        initMultiPointer(gauge, pointerRef, pointer, useCurrentPercent, index);
    }
    var shouldAnimate = (!resize || isFirstAnimation) && pointer.animate !== false;
    if (shouldAnimate) {
        // Mark animation triggered
        if (gauge.multiPointerAnimationTriggered) {
            gauge.multiPointerAnimationTriggered.current[index] = true;
        }
        pointerRef.animationInProgress = true;
        animateMultiPointer(gauge, pointerRef, pointer, index);
    }
    else {
        // Mark animation triggered even when animation is disabled
        if (isFirstAnimation && gauge.multiPointerAnimationTriggered) {
            gauge.multiPointerAnimationTriggered.current[index] = true;
        }
        // Just update position without animation
        updateMultiPointer(pointerRef, pointer, currentPercent, gauge, index);
    }
};
/**
 * Initialize a single pointer element in multi-pointer mode
 */
var initMultiPointer = function (gauge, pointerRef, pointer, useCurrentPercent, index) {
    var _a;
    var _b = pointerRef.context, shouldDrawPath = _b.shouldDrawPath, centerPoint = _b.centerPoint, pointerRadius = _b.pointerRadius, pathLength = _b.pathLength, currentPercent = _b.currentPercent, prevPercent = _b.prevPercent;
    var startPercent = useCurrentPercent ? currentPercent : (prevPercent !== null && prevPercent !== void 0 ? prevPercent : 0);
    // Get color - use pointer color or arc color
    var initialColor = pointer.color || arcHooks.getColorByPercentage(currentPercent, gauge);
    // Remove existing element if any (from ref)
    if (pointerRef.element) {
        pointerRef.element.remove();
    }
    // CRITICAL: Also remove any orphaned DOM elements with same class to prevent duplicates
    // This handles cases where ref was lost but DOM element still exists
    (_a = gauge.g.current) === null || _a === void 0 ? void 0 : _a.selectAll(".multi-pointer-".concat(index)).remove();
    // Create new pointer group
    pointerRef.element = gauge.g.current
        .append("g")
        .attr("class", "multi-pointer multi-pointer-".concat(index));
    if (shouldDrawPath) {
        pointerRef.context.pathStr = calculatePointerPathForConfig(gauge, pointerRef, pointer, startPercent);
        var pathElement = pointerRef.element.append("path")
            .attr("d", pointerRef.context.pathStr)
            .attr("fill", initialColor);
        // Add stroke if configured
        var strokeWidth = pointer.strokeWidth || 0;
        if (strokeWidth > 0) {
            var strokeColor = pointer.strokeColor || 'rgba(255, 255, 255, 0.8)';
            pathElement
                .attr("stroke", strokeColor)
                .attr("stroke-width", strokeWidth)
                .attr("stroke-linejoin", "round");
        }
        pointerRef.path = pathElement;
        // Add grab handle if drag is enabled
        if (gauge.props.onPointerChange && !pointer.hideGrabHandle) {
            var tipPosition = calculatePointerTipPositionForConfig(gauge, pointerRef, pointer, startPercent);
            var handleRadius = Math.max(6, pointerRadius * 0.8);
            pointerRef.element
                .append("circle")
                .attr("class", "pointer-grab-handle pointer-grab-handle-".concat(index))
                .attr("cx", tipPosition.x)
                .attr("cy", tipPosition.y)
                .attr("r", handleRadius)
                .attr("fill", "rgba(255, 255, 255, 0.3)")
                .attr("stroke", initialColor)
                .attr("stroke-width", 2)
                .style("cursor", "grab");
        }
    }
    // Add base circle for needle type
    if (pointer.type === PointerType.Needle) {
        var needleBaseCircle = pointerRef.element
            .append("circle")
            .attr("cx", 0)
            .attr("cy", 0)
            .attr("r", pointerRadius)
            .attr("fill", pointer.baseColor || initialColor);
        var strokeWidth = pointer.strokeWidth || 0;
        if (strokeWidth > 0) {
            var strokeColor = pointer.strokeColor || 'rgba(255, 255, 255, 0.8)';
            needleBaseCircle
                .attr("stroke", strokeColor)
                .attr("stroke-width", strokeWidth);
        }
    }
    else if (pointer.type === PointerType.Blob) {
        var arcColor = arcHooks.getColorByPercentage(currentPercent, gauge);
        var strokeColor = pointer.strokeColor || arcColor;
        var strokeWidth = pointer.strokeWidth !== undefined ? pointer.strokeWidth : 8;
        pointerRef.element
            .append("circle")
            .attr("cx", 0)
            .attr("cy", 0)
            .attr("r", pointerRadius)
            .attr("fill", pointer.baseColor || initialColor)
            .attr("stroke", strokeColor)
            .attr("stroke-width", strokeWidth * pointerRadius / 10);
        pointerRef.context.prevColor = arcColor;
        if (gauge.props.onPointerChange) {
            pointerRef.element.select("circle").style("cursor", "grab");
        }
    }
    // Set initial position
    setMultiPointerPosition(pointerRef, pointer, pointerRadius, startPercent, gauge);
    // Setup drag if enabled
    if (gauge.props.onPointerChange) {
        setupMultiPointerDrag(gauge, pointerRef, pointer, index);
    }
};
/**
 * Animate a single pointer in multi-pointer mode
 */
var animateMultiPointer = function (gauge, pointerRef, pointer, index) {
    var _a;
    var _b = pointerRef.context, prevPercent = _b.prevPercent, currentPercent = _b.currentPercent, prevProgress = _b.prevProgress;
    var isGrafana = gauge.props.type === GaugeType.Grafana;
    var isPrimaryPointer = index === 0;
    var maxFps = (_a = pointer.maxFps) !== null && _a !== void 0 ? _a : 60;
    var minFrameTime = maxFps > 0 ? 1000 / maxFps : 0;
    var lastFrameTime = 0;
    pointerRef.element
        .transition()
        .delay(pointer.animationDelay || 100)
        .ease(pointer.elastic ? easeElastic : easeExpOut)
        .duration(pointer.animationDuration || 3000)
        .tween("progress", function () {
        var currentInterpolatedPercent = interpolateNumber(prevPercent, currentPercent);
        return function (percentOfPercent) {
            var _a;
            var now = performance.now();
            if (now - lastFrameTime < minFrameTime)
                return;
            lastFrameTime = now;
            var progress = currentInterpolatedPercent(percentOfPercent);
            var threshold = (_a = pointer.animationThreshold) !== null && _a !== void 0 ? _a : 0.001;
            if (Math.abs(progress - pointerRef.context.prevProgress) >= threshold || percentOfPercent >= 1) {
                updateMultiPointer(pointerRef, pointer, progress, gauge, index);
                pointerRef.context.prevProgress = progress;
                // For Grafana, animate arc fill using primary pointer's progress
                if (isGrafana && isPrimaryPointer) {
                    arcHooks.updateGrafanaArc(gauge, progress);
                }
            }
        };
    })
        .on("end", function () {
        var _a, _b, _c;
        pointerRef.animationInProgress = false;
        // Check if all animations are done
        var allDone = (_b = (_a = gauge.multiPointers) === null || _a === void 0 ? void 0 : _a.current.every(function (p) { return !(p === null || p === void 0 ? void 0 : p.animationInProgress); })) !== null && _b !== void 0 ? _b : true;
        // Apply pending config changes when all multi-pointer animations complete
        if (allDone && ((_c = gauge.pendingConfigChange) === null || _c === void 0 ? void 0 : _c.current)) {
            gauge.pendingConfigChange.current = false;
            requestAnimationFrame(function () {
                var chartHooks = require('./chart');
                chartHooks.initChart(gauge, false);
            });
        }
    });
};
/**
 * Update a single pointer position in multi-pointer mode
 */
var updateMultiPointer = function (pointerRef, pointer, percentage, gauge, index) {
    var _a, _b, _c, _d;
    var _e = pointerRef.context, pointerRadius = _e.pointerRadius, shouldDrawPath = _e.shouldDrawPath, prevColor = _e.prevColor;
    setMultiPointerPosition(pointerRef, pointer, pointerRadius, percentage, gauge);
    if (shouldDrawPath) {
        pointerRef.context.pathStr = calculatePointerPathForConfig(gauge, pointerRef, pointer, percentage);
        (_a = pointerRef.path) === null || _a === void 0 ? void 0 : _a.attr("d", pointerRef.context.pathStr);
        // Update grab handle position
        var grabHandle = (_b = pointerRef.element) === null || _b === void 0 ? void 0 : _b.select(".pointer-grab-handle-".concat(index));
        if (grabHandle && !grabHandle.empty()) {
            var tipPosition = calculatePointerTipPositionForConfig(gauge, pointerRef, pointer, percentage);
            grabHandle.attr("cx", tipPosition.x).attr("cy", tipPosition.y);
        }
        // Update color if not fixed
        if (!pointer.color) {
            var newColor = arcHooks.getColorByPercentage(percentage, gauge);
            (_c = pointerRef.path) === null || _c === void 0 ? void 0 : _c.attr("fill", newColor);
            grabHandle === null || grabHandle === void 0 ? void 0 : grabHandle.attr("stroke", newColor);
        }
    }
    // Update blob stroke color
    if (pointer.type === PointerType.Blob && !pointer.strokeColor) {
        var newColor = arcHooks.getColorByPercentage(percentage, gauge);
        if (newColor !== prevColor) {
            (_d = pointerRef.element) === null || _d === void 0 ? void 0 : _d.select("circle").attr("stroke", newColor);
            pointerRef.context.prevColor = newColor;
        }
    }
};
/**
 * Set position for a multi-pointer element
 * Uses the EXACT same positioning logic as single-pointer mode
 */
var setMultiPointerPosition = function (pointerRef, pointer, pointerRadius, percent, gauge) {
    var _a, _b, _c, _d, _e;
    var _f = getEffectiveAngles(gauge), startAngle = _f.startAngle, endAngle = _f.endAngle;
    var angle = startAngle + percent * (endAngle - startAngle);
    var innerR = gauge.dimensions.current.innerRadius;
    var outerR = gauge.dimensions.current.outerRadius;
    if (pointer.type === PointerType.Blob) {
        // SAME as single pointer: Position blob based on blobOffset (0 = inner edge, 0.5 = center, 1 = outer edge)
        var blobOffset = (_a = pointer.blobOffset) !== null && _a !== void 0 ? _a : 0.5;
        var targetRadius = innerR + (outerR - innerR) * blobOffset;
        var x = targetRadius * Math.sin(angle);
        var y = -targetRadius * Math.cos(angle);
        (_b = pointerRef.element) === null || _b === void 0 ? void 0 : _b.attr("transform", "translate(".concat(x, ", ").concat(y, ")"));
    }
    else if (pointer.type === PointerType.Arrow) {
        // SAME as single pointer: Arrow offset is relative to inner radius (not interpolated)
        var arrowOffset = (_c = pointer.arrowOffset) !== null && _c !== void 0 ? _c : 0.72;
        var targetRadius = innerR * arrowOffset;
        var x = targetRadius * Math.sin(angle);
        var y = -targetRadius * Math.cos(angle);
        (_d = pointerRef.element) === null || _d === void 0 ? void 0 : _d.attr("transform", "translate(".concat(x, ", ").concat(y, ")"));
    }
    else {
        // Needle - positioned at center (0, 0)
        (_e = pointerRef.element) === null || _e === void 0 ? void 0 : _e.attr("transform", "translate(0, 0)");
    }
};
/**
 * Get pointer radius for a specific pointer config
 * Uses the EXACT same formula as single-pointer mode
 */
var getPointerRadiusForConfig = function (gauge, pointer) {
    var _a;
    var pointerWidth = (_a = pointer.width) !== null && _a !== void 0 ? _a : 15;
    return pointerWidth * (gauge.dimensions.current.width / 500);
};
/**
 * Calculate pointer path for a specific pointer config
 * Uses the EXACT same logic as single-pointer calculatePointerPath
 */
var calculatePointerPathForConfig = function (gauge, pointerRef, pointer, percent) {
    var _a = pointerRef.context, pointerRadius = _a.pointerRadius, pathLength = _a.pathLength;
    var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
    var d3Angle = startAngle + percent * (endAngle - startAngle);
    // SAME as single pointer calculatePointerPath
    // Calculate needle tip position using D3 angle convention (same as arc)
    var tipX = pathLength * Math.sin(d3Angle);
    var tipY = -pathLength * Math.cos(d3Angle);
    // Calculate base points perpendicular to the needle direction
    var perpAngle = d3Angle + Math.PI / 2;
    var baseOffset = pointerRadius;
    var leftX = baseOffset * Math.sin(perpAngle);
    var leftY = -baseOffset * Math.cos(perpAngle);
    var rightX = -baseOffset * Math.sin(perpAngle);
    var rightY = baseOffset * Math.cos(perpAngle);
    return "M ".concat(leftX, " ").concat(leftY, " L ").concat(tipX, " ").concat(tipY, " L ").concat(rightX, " ").concat(rightY);
};
/**
 * Calculate pointer tip position for a specific pointer config
 * Uses the EXACT same logic as single-pointer calculatePointerTipPosition
 */
var calculatePointerTipPositionForConfig = function (gauge, pointerRef, pointer, percent) {
    var _a;
    var pathLength = pointerRef.context.pathLength;
    var innerR = gauge.dimensions.current.innerRadius;
    var _b = getEffectiveAngles(gauge), startAngle = _b.startAngle, endAngle = _b.endAngle;
    var d3Angle = startAngle + percent * (endAngle - startAngle);
    // For Arrow type, calculate position based on arrow offset + local tip
    if (pointer.type === PointerType.Arrow) {
        var arrowOffset = (_a = pointer.arrowOffset) !== null && _a !== void 0 ? _a : 0.72;
        var targetRadius = innerR * arrowOffset;
        // Arrow translation position
        var transX = targetRadius * Math.sin(d3Angle);
        var transY = -targetRadius * Math.cos(d3Angle);
        // Local tip position (arrow path tip)
        var localTipX = pathLength * Math.sin(d3Angle);
        var localTipY = -pathLength * Math.cos(d3Angle);
        return {
            x: transX + localTipX,
            y: transY + localTipY,
        };
    }
    // For Needle type, tip is at pathLength from center along the angle
    return {
        x: pathLength * Math.sin(d3Angle),
        y: -pathLength * Math.cos(d3Angle),
    };
};
/**
 * Setup drag behavior for a multi-pointer
 */
var setupMultiPointerDrag = function (gauge, pointerRef, pointer, index) {
    var _a, _b;
    var onPointerChange = gauge.props.onPointerChange;
    if (!onPointerChange)
        return;
    var dragBehavior = drag()
        .on("start", function () {
        select(this).style("cursor", "grabbing");
    })
        .on("drag", function (event) {
        var value = getValueFromPosition(gauge, event.sourceEvent.clientX, event.sourceEvent.clientY);
        onPointerChange(index, value);
    })
        .on("end", function () {
        select(this).style("cursor", "grab");
    });
    // Apply drag to the pointer element
    (_a = pointerRef.element) === null || _a === void 0 ? void 0 : _a.call(dragBehavior);
    // Apply drag to grab handle
    var grabHandle = (_b = pointerRef.element) === null || _b === void 0 ? void 0 : _b.select(".pointer-grab-handle-".concat(index));
    if (grabHandle && !grabHandle.empty()) {
        grabHandle.call(dragBehavior);
    }
};
/**
 * Clear all multi-pointer elements
 */
export var clearMultiPointers = function (gauge) {
    var _a, _b, _c;
    (_a = gauge.g.current) === null || _a === void 0 ? void 0 : _a.selectAll(".multi-pointer").remove();
    if ((_b = gauge.multiPointers) === null || _b === void 0 ? void 0 : _b.current) {
        gauge.multiPointers.current = [];
    }
    if ((_c = gauge.multiPointerAnimationTriggered) === null || _c === void 0 ? void 0 : _c.current) {
        gauge.multiPointerAnimationTriggered.current = [];
    }
};
