// react-gauge-component/dist/lib/GaugeComponent/hooks/utils.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
export var calculatePercentage = function (minValue, maxValue, value) {
    if (value < minValue) {
        return 0;
    }
    else if (value > maxValue) {
        return 1;
    }
    else {
        var percentage = (value - minValue) / (maxValue - minValue);
        return (percentage);
    }
};
/**
 * Convert a percentage (0-1) back to a value within the min/max range
 */
export var percentToValue = function (percent, minValue, maxValue) {
    return minValue + percent * (maxValue - minValue);
};
export var isEmptyObject = function (obj) {
    return Object.keys(obj).length === 0 && obj.constructor === Object;
};
export var mergeObjects = function (obj1, obj2) {
    var mergedObj = __assign({}, obj1);
    Object.keys(obj2).forEach(function (key) {
        var val1 = obj1[key];
        var val2 = obj2[key];
        if (Array.isArray(val1) && Array.isArray(val2)) {
            mergedObj[key] = val2;
        }
        else if (typeof val1 === 'object' && typeof val2 === 'object') {
            mergedObj[key] = mergeObjects(val1, val2);
        }
        else if (val2 !== undefined) {
            mergedObj[key] = val2;
        }
    });
    return mergedObj;
};
//Returns the angle (in rad) for the given 'percent' value where percent = 1 means 100% and is 180 degree angle
export var percentToRad = function (percent, angle) {
    return percent * (Math.PI / angle);
};
export var floatingNumber = function (value, maxDigits) {
    if (maxDigits === void 0) { maxDigits = 2; }
    return Math.round(value * Math.pow(10, maxDigits)) / Math.pow(10, maxDigits);
};
// Function to normalize a value between a new min and max
export function normalize(value, min, max) {
    return ((value - min) / (max - min)) * 100;
}
export var degToRad = function (degrees) {
    return degrees * (Math.PI / 180);
};
export var radToDeg = function (radians) {
    return radians * (180 / Math.PI);
};
export var getCurrentGaugePercentageByValue = function (value, gauge) { return calculatePercentage(gauge.minValue, gauge.maxValue, value); };
export var getCurrentGaugeValueByPercentage = function (percentage, gauge) {
    var minValue = gauge.props.minValue;
    var maxValue = gauge.props.maxValue;
    var value = minValue + (percentage) * (maxValue - minValue);
    return value;
};
export var camelCaseToKebabCase = function (str) { return str.replace(/[A-Z]/g, function (letter) { return "-".concat(letter.toLowerCase()); }); };
/**
 * Shallow comparison of two values. More performant than JSON.stringify for prop comparisons.
 * Handles primitives, arrays (shallow), and objects (shallow, one level deep for nested objects).
 */
export var shallowEqual = function (a, b) {
    if (a === b)
        return true;
    if (a == null || b == null)
        return a === b;
    if (typeof a !== typeof b)
        return false;
    if (typeof a !== 'object')
        return a === b;
    if (Array.isArray(a) !== Array.isArray(b))
        return false;
    if (Array.isArray(a)) {
        if (a.length !== b.length)
            return false;
        for (var i = 0; i < a.length; i++) {
            if (!shallowEqualSimple(a[i], b[i]))
                return false;
        }
        return true;
    }
    var keysA = Object.keys(a);
    var keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (var _i = 0, keysA_1 = keysA; _i < keysA_1.length; _i++) {
        var key = keysA_1[_i];
        if (!Object.prototype.hasOwnProperty.call(b, key))
            return false;
        if (!shallowEqualSimple(a[key], b[key]))
            return false;
    }
    return true;
};
/**
 * Simple shallow comparison for nested values (one level only)
 */
var shallowEqualSimple = function (a, b) {
    if (a === b)
        return true;
    if (a == null || b == null)
        return a === b;
    if (typeof a !== typeof b)
        return false;
    if (typeof a !== 'object')
        return a === b;
    if (Array.isArray(a) !== Array.isArray(b))
        return false;
    if (Array.isArray(a)) {
        if (a.length !== b.length)
            return false;
        for (var i = 0; i < a.length; i++) {
            if (a[i] !== b[i])
                return false;
        }
        return true;
    }
    var keysA = Object.keys(a);
    var keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (var _i = 0, keysA_2 = keysA; _i < keysA_2.length; _i++) {
        var key = keysA_2[_i];
        if (a[key] !== b[key])
            return false;
    }
    return true;
};
